<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.189">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="J. Serizay &amp; O. Ashenberg">
<title>Single-cell RNAseq analysis with R/Bioconductor - 9&nbsp; Lab 5: Dimension reduction, clustering and annotation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../content/day3/Lecture5_batchcorrection.html" rel="next">
<link href="../../content/day3/Lecture4_clustering.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }"><div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">
<span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Lab 5: Dimension reduction, clustering and annotation</span>
</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">Single-cell RNAseq analysis with R/Bioconductor</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/js2264/scRNAseq_Physalia_2023/" title="Source Code" class="sidebar-tool px-1"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">Welcome</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/program.html" class="sidebar-item-text sidebar-link">Program</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/Rstudio.html" class="sidebar-item-text sidebar-link">RStudio</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/prerequisites.html" class="sidebar-item-text sidebar-link">Prerequisites</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Day 1</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lecture1_introduction.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Lecture 1 - Introduction to scRNAseq analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lab1_Intro.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Lab 1: Familiarizing yourself with the course AWS instance</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lecture2_Bcl2matrix.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Lecture 2 - From sequencing reads to expression matrices</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lab2_processingreads.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Lab 2: From .bcl to count matrix</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">Day 2</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lab3_Rbioc.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Lab 3: Introduction to R/Bioconductor</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lecture3_qualitycontrol.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Lecture 3 - Quality control for scRNA-Seq data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lab4_data_wrangling_scRNAseq.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Lab 4 - Single-cell RNA-seq data wrangling</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">Day 3</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lecture4_clustering.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Lecture 4 - Identifying cell populations</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lab5_clustering.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Lab 5: Dimension reduction, clustering and annotation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lecture5_batchcorrection.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Lecture 5 - Data integration and batch effect correction</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lab6_batch_correction.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Lab 6 - Batch correction</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">Day 4</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lecture6_pseudotime.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Lecture 6 - Trajectories and pseudotimes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lab7_pseudotime.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Lab 7: Pseudotime analyses</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lecture7_ATAC.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Lecture 7 - Advances in single-cell genomics: the epigenome</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lab8_atac-seq.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Lab 8 - Single-cell ATAC-seq analysis workflow</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">Day 5</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day5/Lecture8_spatial-transcriptomics.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Lecture 8 - Advances in single-cell genomics: spatial transcriptomics</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/extra.html" class="sidebar-item-text sidebar-link">Extra resources</a>
  </div>
</li>
    </ul>
</div>
</nav><!-- margin-sidebar --><div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#dimensional-reduction-for-clustering" id="toc-dimensional-reduction-for-clustering" class="nav-link active" data-scroll-target="#dimensional-reduction-for-clustering"><span class="toc-section-number">9.1</span>  Dimensional reduction for clustering</a>
  <ul class="collapse">
<li><a href="#preparing-dataset" id="toc-preparing-dataset" class="nav-link" data-scroll-target="#preparing-dataset"><span class="toc-section-number">9.1.1</span>  Preparing dataset</a></li>
  <li><a href="#normalize-counts-using-scran" id="toc-normalize-counts-using-scran" class="nav-link" data-scroll-target="#normalize-counts-using-scran"><span class="toc-section-number">9.1.2</span>  Normalize counts using <code>scran</code></a></li>
  <li><a href="#feature-selection" id="toc-feature-selection" class="nav-link" data-scroll-target="#feature-selection"><span class="toc-section-number">9.1.3</span>  Feature selection</a></li>
  <li><a href="#pca-on-filtered-dataset" id="toc-pca-on-filtered-dataset" class="nav-link" data-scroll-target="#pca-on-filtered-dataset"><span class="toc-section-number">9.1.4</span>  PCA on filtered dataset</a></li>
  </ul>
</li>
  <li>
<a href="#clustering" id="toc-clustering" class="nav-link" data-scroll-target="#clustering"><span class="toc-section-number">9.2</span>  Clustering</a>
  <ul class="collapse">
<li><a href="#clustering-algorithms" id="toc-clustering-algorithms" class="nav-link" data-scroll-target="#clustering-algorithms"><span class="toc-section-number">9.2.1</span>  Clustering algorithms</a></li>
  <li><a href="#dimensional-reduction-for-clustering-visualization" id="toc-dimensional-reduction-for-clustering-visualization" class="nav-link" data-scroll-target="#dimensional-reduction-for-clustering-visualization"><span class="toc-section-number">9.2.2</span>  Dimensional reduction for clustering visualization</a></li>
  <li><a href="#for-the-pros-of-clustering-compare-different-clustering-approaches" id="toc-for-the-pros-of-clustering-compare-different-clustering-approaches" class="nav-link" data-scroll-target="#for-the-pros-of-clustering-compare-different-clustering-approaches"><span class="toc-section-number">9.2.3</span>  For the pros of clustering… Compare different clustering approaches</a></li>
  </ul>
</li>
  <li>
<a href="#cell-annotation" id="toc-cell-annotation" class="nav-link" data-scroll-target="#cell-annotation"><span class="toc-section-number">9.3</span>  Cell annotation</a>
  <ul class="collapse">
<li><a href="#find-marker-genes" id="toc-find-marker-genes" class="nav-link" data-scroll-target="#find-marker-genes"><span class="toc-section-number">9.3.1</span>  Find marker genes</a></li>
  <li><a href="#automated-cell-annotation" id="toc-automated-cell-annotation" class="nav-link" data-scroll-target="#automated-cell-annotation"><span class="toc-section-number">9.3.2</span>  Automated cell annotation</a></li>
  </ul>
</li>
  <li><a href="#bonus" id="toc-bonus" class="nav-link" data-scroll-target="#bonus"><span class="toc-section-number">9.4</span>  Bonus</a></li>
  <li><a href="#acknowledgements" id="toc-acknowledgements" class="nav-link" data-scroll-target="#acknowledgements"><span class="toc-section-number">9.5</span>  Acknowledgements</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info"><span class="toc-section-number">9.6</span>  Session info</a></li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/js2264/scRNAseq_Physalia_2023/edit/main/content/day3/Lab5_clustering.qmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/js2264/scRNAseq_Physalia_2023/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title d-none d-lg-block">
<span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Lab 5: Dimension reduction, clustering and annotation</span>
</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header><div class="cell">

</div>
<section id="dimensional-reduction-for-clustering" class="level2" data-number="9.1"><h2 data-number="9.1" class="anchored" data-anchor-id="dimensional-reduction-for-clustering">
<span class="header-section-number">9.1</span> Dimensional reduction for clustering</h2>
<section id="preparing-dataset" class="level3" data-number="9.1.1"><h3 data-number="9.1.1" class="anchored" data-anchor-id="preparing-dataset">
<span class="header-section-number">9.1.1</span> Preparing dataset</h3>
<p>We will prepare scRNAseq data from a PBMC run, provided by 10X and hosted by <code>Bioconductor</code> as a package.</p>
<ul>
<li><strong>Which package from <code>Bioconductor</code> gives streamlined access to PBMC scRNAseq dataset from 10X Genomics?</strong></li>
<li><strong>What does the object contain (type of data, number of cells, batches, organism, …)? Can you get the same data from somewhere else?</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">TENxPBMCData</span><span class="fu">::</span><span class="fu">TENxPBMCData</span><span class="op">(</span><span class="st">'pbmc4k'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">scuttle</span><span class="fu">::</span><span class="fu">uniquifyFeatureNames</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">ENSEMBL_ID</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol_TENx</span><span class="op">)</span></span>
<span><span class="va">sce</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">Library</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="normalize-counts-using-scran" class="level3" data-number="9.1.2"><h3 data-number="9.1.2" class="anchored" data-anchor-id="normalize-counts-using-scran">
<span class="header-section-number">9.1.2</span> Normalize counts using <code>scran</code>
</h3>
<p>Just like in bulk high-throughput sequencing experiments, scRNAseq counts have to be normalized to the sequencing depth for each cell. We can define the library size as the total sum of counts across all genes for each cell, the expected value of which is assumed to scale with any cell-specific biases. However, this relies on the assumption that within the entire dataset, most genes are non-differentially expressed and expressed roughly within the same range. Depending on the set up of the scRNAseq experiment, this can be entirely false. To avoid relying on this hypothesis, we can (1) quickly pre-cluster cells, then (2) normalize cells using their library size factor separately in each cluster, then (3) rescaling size factors so that they are comparable across clusters.</p>
<p>All of this can be done very simply using the combo <code>quickCluster() + computeSumFactors() + logNormCounts()</code> from <code>scran/scuttle</code> packages.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu">quickCluster</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu">computeSumFactors</span><span class="op">(</span><span class="va">sce</span>, cluster <span class="op">=</span> <span class="va">clusters</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scuttle</span><span class="fu">::</span><span class="fu">logNormCounts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assays</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="feature-selection" class="level3" data-number="9.1.3"><h3 data-number="9.1.3" class="anchored" data-anchor-id="feature-selection">
<span class="header-section-number">9.1.3</span> Feature selection</h3>
<p>We often use scRNAseq data in exploratory analyses to characterize heterogeneity across cells. Procedures like clustering and dimensionality reduction compare cells based on their gene expression profiles. The choice of genes to include in this comparison may have a major impact on the performance of downstream methods. Ideally, one wants to only select genes that contain useful information about the biology of the system while removing genes that contain random noise. This aims to preserve interesting biological structure without the variance that obscures that structure.</p>
<p>The simplest approach to feature selection is to compute the variance of the log-normalized expression values, to select the most variable genes. Modelling of the mean-variance relationship can be achieved by the <code>modelGeneVar()</code> function from the <code>scran</code> package.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sce_filtered_variance</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu">modelGeneVar</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">HVGs</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu">getTopHVGs</span><span class="op">(</span><span class="va">sce_filtered_variance</span>, prop <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">isHVG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">HVGs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">isHVG</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## --- Visualizing the mean-variance fit</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>    mean <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">sce_filtered_variance</span><span class="op">)</span><span class="op">$</span><span class="va">mean</span>, </span>
<span>    var <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">sce_filtered_variance</span><span class="op">)</span><span class="op">$</span><span class="va">var</span>, </span>
<span>    trend <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">sce_filtered_variance</span><span class="op">)</span><span class="op">$</span><span class="fu">trend</span><span class="op">(</span><span class="va">mean</span><span class="op">)</span>, </span>
<span>    HVG <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">isHVG</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">mean</span>, y <span class="op">=</span> <span class="va">var</span>, col <span class="op">=</span> <span class="va">HVG</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">mean</span>, y <span class="op">=</span> <span class="va">trend</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">'darkred'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">'Gene mean exp. (norm.)'</span>, y <span class="op">=</span> <span class="st">'Gene exp. variance'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="pca-on-filtered-dataset" class="level3" data-number="9.1.4"><h3 data-number="9.1.4" class="anchored" data-anchor-id="pca-on-filtered-dataset">
<span class="header-section-number">9.1.4</span> PCA on filtered dataset</h3>
<p>We now have normalized counts filtered for the top 500 genes varying with the greatest biological significance.<br>
Still, that represents a 500 x nCells (~8,000) dataset (each row being a feature). This is still too big to reliably use in standard clustering approaches. We can further compress the dataset. The most widely used approach is <code>PCA</code>: it computes a small number of “components” (typically 5-50) optimally summarizing the variability of the whole dataset, while retaining linearity of the underlying numerical data and being computationallt quite efficient.</p>
<ul>
<li><strong>Leverage <code>scater</code> package to compute a <code>PCA</code> embedding of the filtered data by taking into account the technical variability.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu">denoisePCA</span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, </span>
<span>    technical <span class="op">=</span> <span class="va">sce_filtered_variance</span>, </span>
<span>    subset.row <span class="op">=</span> <span class="va">HVGs</span>, </span>
<span>    min.rank <span class="op">=</span> <span class="fl">15</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">'PCA'</span>, colour_by <span class="op">=</span> <span class="st">'sizeFactor'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'denoised PCA'</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section></section><section id="clustering" class="level2" data-number="9.2"><h2 data-number="9.2" class="anchored" data-anchor-id="clustering">
<span class="header-section-number">9.2</span> Clustering</h2>
<p>Clustering is an unsupervised learning procedure that is used in scRNA-seq data analysis to empirically define groups of cells with similar expression profiles. Its primary purpose is to summarize the data in a digestible format for human interpretation.</p>
<p>After annotation based on marker genes, the clusters can be treated as proxies for more abstract biological concepts such as cell types or states. Clustering is thus a critical step for extracting biological insights from scRNA-seq data.</p>
<section id="clustering-algorithms" class="level3" data-number="9.2.1"><h3 data-number="9.2.1" class="anchored" data-anchor-id="clustering-algorithms">
<span class="header-section-number">9.2.1</span> Clustering algorithms</h3>
<p>Three main approaches can be used:</p>
<ol type="1">
<li>Hierarchical clustering</li>
<li>k-means clustering</li>
<li>Graph-based clustering</li>
</ol>
<p>Today, we will focus on graph-based clustering, as it is becoming the standard for scRNAseq: it is a flexible and scalable technique for clustering even the largest scRNA-seq datasets. We first build a graph where each node is a cell that is connected by edges to its nearest neighbors in the high-dimensional space. Edges are weighted based on the similarity between the cells involved, with higher weight given to cells that are more closely related.</p>
<ul>
<li><strong>Compute graph-based clustering of the PBMC dataset.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu">buildSNNGraph</span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, </span>
<span>    k <span class="op">=</span> <span class="fl">5</span>, </span>
<span>    use.dimred <span class="op">=</span> <span class="st">'PCA'</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sce_clust</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu">cluster_louvain</span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">membership</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">sce_clust</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">clusters_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">sce_clust</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>What are the main parameters to choose? How do they impact the clustering?</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">graph2</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu">buildSNNGraph</span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, </span>
<span>    k <span class="op">=</span> <span class="fl">50</span>, </span>
<span>    use.dimred <span class="op">=</span> <span class="st">'PCA'</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sce_clust2</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu">cluster_louvain</span><span class="op">(</span><span class="va">graph2</span><span class="op">)</span><span class="op">$</span><span class="va">membership</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">sce_clust</span>, <span class="va">sce_clust2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="dimensional-reduction-for-clustering-visualization" class="level3" data-number="9.2.2"><h3 data-number="9.2.2" class="anchored" data-anchor-id="dimensional-reduction-for-clustering-visualization">
<span class="header-section-number">9.2.2</span> Dimensional reduction for clustering visualization</h3>
<p><code>PCA</code> is a powerful linear approach to compress large datasets into smaller dimensional spaces. However, it struggles at emphasizing the existence of clusters in complex datasets, when visualized in 2D.</p>
<p><code>scater</code> provides a handy way to perform more complex data embeddings:</p>
<pre><code>- tSNE
- UMAP
- Diffusion Map
- Multi-Dimensional Scaling (MDS)
- Non-negative Matrix Factorization (NMF)</code></pre>
<ul>
<li><strong>Explore the different dimensional reduction algorithms, trying different hyperparameters combinations.</strong></li>
</ul>
<p>When you run these commands, pay attention to how long each command takes to run! While this run, check the <code>Help</code> page for each function (e.g.&nbsp;<code>?runTSNE</code>)</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDims</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu">runTSNE</span><span class="op">(</span><span class="va">sce</span>, subset_row <span class="op">=</span> <span class="va">HVGs</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu">runUMAP</span><span class="op">(</span><span class="va">sce</span>, subset_row <span class="op">=</span> <span class="va">HVGs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDims</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'UMAP'</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>Use the <code>scater::plotReducedDim()</code> function to plot cells in each embedding. Comment.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">p</span><span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">'PCA'</span>, colour_by <span class="op">=</span> <span class="st">'clusters_graph'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'denoised PCA'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scater</span><span class="fu">::</span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">'TSNE'</span>, colour_by <span class="op">=</span> <span class="st">'clusters_graph'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'tSNE'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scater</span><span class="fu">::</span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">'UMAP'</span>, colour_by <span class="op">=</span> <span class="st">'clusters_graph'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'UMAP'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="for-the-pros-of-clustering-compare-different-clustering-approaches" class="level3" data-number="9.2.3"><h3 data-number="9.2.3" class="anchored" data-anchor-id="for-the-pros-of-clustering-compare-different-clustering-approaches">
<span class="header-section-number">9.2.3</span> For the pros of clustering… Compare different clustering approaches</h3>
<p>Leveraging the <code>bluster</code> package, different clustering approaches can be performed using a uniformed syntax, to compare their output.</p>
<ul>
<li><strong>Using <code>clusterSweep()</code>, compare the effect of different <code>k</code> neighbor values when performing graph-based clustering.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu">bluster</span><span class="fu">::</span><span class="fu">clusterSweep</span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">'PCA'</span><span class="op">)</span>, </span>
<span>    BLUSPARAM <span class="op">=</span> <span class="fu">bluster</span><span class="fu">::</span><span class="fu">SNNGraphParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>    k <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5L</span>, <span class="fl">15L</span>, <span class="fl">25L</span>, <span class="fl">50L</span><span class="op">)</span>, </span>
<span>    cluster.fun <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"louvain"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span></span>
<span><span class="va">clusters</span><span class="op">$</span><span class="va">parameters</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggraph.data-imaginist.com">ggraph</a></span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span></span>
<span>    <span class="fu">clustree</span><span class="fu">::</span><span class="fu">clustree</span><span class="op">(</span></span>
<span>        <span class="va">clusters</span><span class="op">$</span><span class="va">clusters</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        prefix <span class="op">=</span> <span class="st">'X'</span>,</span>
<span>        edge_arrow<span class="op">=</span><span class="cn">FALSE</span></span>
<span>    <span class="op">)</span>, </span>
<span>    <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span></span>
<span>        <span class="fu">scater</span><span class="fu">::</span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">'TSNE'</span>, colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/AsIs.html">I</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">$</span><span class="va">clusters</span><span class="op">[</span>, <span class="st">'k.5_cluster.fun.louvain'</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'k = 5'</span><span class="op">)</span>,</span>
<span>        <span class="fu">scater</span><span class="fu">::</span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">'TSNE'</span>, colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/AsIs.html">I</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">$</span><span class="va">clusters</span><span class="op">[</span>, <span class="st">'k.15_cluster.fun.louvain'</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'k = 15'</span><span class="op">)</span>,</span>
<span>        <span class="fu">scater</span><span class="fu">::</span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">'TSNE'</span>, colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/AsIs.html">I</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">$</span><span class="va">clusters</span><span class="op">[</span>, <span class="st">'k.25_cluster.fun.louvain'</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'k = 25'</span><span class="op">)</span>,</span>
<span>        <span class="fu">scater</span><span class="fu">::</span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">'TSNE'</span>, colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/AsIs.html">I</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">$</span><span class="va">clusters</span><span class="op">[</span>, <span class="st">'k.50_cluster.fun.louvain'</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'k = 50'</span><span class="op">)</span></span>
<span>    <span class="op">)</span>, </span>
<span>    nrow <span class="op">=</span> <span class="fl">2</span>, </span>
<span>    rel_heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.7</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">$</span><span class="va">clusters</span><span class="op">[</span>, <span class="st">'k.5_cluster.fun.louvain'</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section></section><section id="cell-annotation" class="level2" data-number="9.3"><h2 data-number="9.3" class="anchored" data-anchor-id="cell-annotation">
<span class="header-section-number">9.3</span> Cell annotation</h2>
<section id="find-marker-genes" class="level3" data-number="9.3.1"><h3 data-number="9.3.1" class="anchored" data-anchor-id="find-marker-genes">
<span class="header-section-number">9.3.1</span> Find marker genes</h3>
<p>To interpret clustering results, one needs to identify the genes that drive separation between clusters. These marker genes allow to assign biological meaning to each cluster based on their functional annotation. In the most obvious case, the marker genes for each cluster are <em>a priori</em> associated with particular cell types, allowing us to treat the clustering as a <em>proxy</em> for cell type identity.</p>
<p>A general strategy is to perform DE tests between pairs of clusters and then combine results into a single ranking of marker genes for each cluster.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu">findMarkers</span><span class="op">(</span><span class="va">sce</span>, groups <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">clusters_graph</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>Find markers strongly overexpressed in each cluster. Check <code>?scran::findMarkers</code> to find the right options to use.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu">findMarkers</span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, </span>
<span>    groups <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">clusters_graph</span>, </span>
<span>    direction <span class="op">=</span> <span class="st">"up"</span>, </span>
<span>    lfc <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">markers</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">markers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">df</span><span class="op">[</span><span class="va">df</span><span class="op">$</span><span class="va">Top</span> <span class="op">&lt;=</span> <span class="fl">5</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>Plot average expression of the first marker of the first cluster in UMAP</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">'TSNE'</span>, colour_by <span class="op">=</span> <span class="va">markers</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="automated-cell-annotation" class="level3" data-number="9.3.2"><h3 data-number="9.3.2" class="anchored" data-anchor-id="automated-cell-annotation">
<span class="header-section-number">9.3.2</span> Automated cell annotation</h3>
<p>Many cell type reference databases are available over the Internet. Today, we will use a reference constructed from <code>Blueprint</code> and <code>ENCODE</code> data (<code>Martens and Stunnenberg 2013</code>; <code>The ENCODE Project Consortium 2012</code>). This reference is available as a <code>SummarizedExperiment</code> containing log-normalized gene expression for manually annotated samples.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu">celldex</span><span class="fu">::</span><span class="fu">BlueprintEncodeData</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">prediction_types</span> <span class="op">&lt;-</span> <span class="fu">SingleR</span><span class="fu">::</span><span class="fu">SingleR</span><span class="op">(</span></span>
<span>    test <span class="op">=</span> <span class="va">sce</span>, </span>
<span>    ref <span class="op">=</span> <span class="va">ref</span>, </span>
<span>    labels <span class="op">=</span> <span class="va">ref</span><span class="op">$</span><span class="va">label.main</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="va">prediction_types</span><span class="op">$</span><span class="va">labels</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">annotation</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">annotation</span>, <span class="va">sce</span><span class="op">$</span><span class="va">clusters_graph</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>Using <code>scater</code> and <code>SingleR</code> utilities, visually compare the annotation scores for cells in each cluster.</strong></li>
<li><strong>Did the automated annotation work robuslty? How does it compare to our clustering? Is automated annotation as sensitive as graph-based clustering?</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">SingleR</span><span class="fu">::</span><span class="fu">plotScoreHeatmap</span><span class="op">(</span><span class="va">prediction_types</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">'TSNE'</span>, colour_by <span class="op">=</span> <span class="st">'annotation'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'Automated annotation'</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">pheatmap</span><span class="fu">::</span><span class="fu">pheatmap</span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span>Annotation <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">annotation</span>, Cluster <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">clusters_graph</span><span class="op">)</span><span class="op">+</span><span class="fl">10</span><span class="op">)</span>, </span>
<span>    color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"darkred"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">101</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section></section><section id="bonus" class="level2" data-number="9.4"><h2 data-number="9.4" class="anchored" data-anchor-id="bonus">
<span class="header-section-number">9.4</span> Bonus</h2>
<p>Try to fill in the analysis template in <code>bin/prepare_Ernst.R</code> to execute the different processing/analysis steps we covered in the previous exercises and this one. If you prefer using <code>Seurat</code>, don’t hesitate to modify the base template!</p>
</section><section id="acknowledgements" class="level2" data-number="9.5"><h2 data-number="9.5" class="anchored" data-anchor-id="acknowledgements">
<span class="header-section-number">9.5</span> Acknowledgements</h2>
<p>This exercise was adapted from Chapts. 7-12 of <a href="https://bioconductor.org/books/release/OSCA/">Orchestrating Single-Cell Analysis with Bioconductor</a>.</p>
</section><section id="session-info" class="level2" data-number="9.6"><h2 data-number="9.6" class="anchored" data-anchor-id="session-info">
<span class="header-section-number">9.6</span> Session info</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.0 (2023-04-21)
 os       macOS Monterey 12.5.1
 system   aarch64, darwin20
 ui       X11
 language (EN)
 collate  en_GB.UTF-8
 ctype    en_GB.UTF-8
 tz       Europe/Paris
 date     2023-06-05
 pandoc   2.19.2 @ /opt/homebrew/bin/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 package                * version   date (UTC) lib source
 AnnotationDbi          * 1.61.0    2022-11-01 [1] Bioconductor
 AnnotationHub          * 3.7.3     2023-03-01 [1] Bioconductor
 Biobase                * 2.59.0    2022-11-01 [1] Bioconductor
 BiocFileCache          * 2.7.2     2023-02-17 [1] Bioconductor
 BiocGenerics           * 0.45.0    2022-11-01 [1] Bioconductor
 BiocManager            * 1.30.20   2023-02-24 [1] CRAN (R 4.3.0)
 BiocVersion              3.17.1    2022-12-20 [1] Bioconductor
 Biostrings               2.67.0    2022-11-01 [1] Bioconductor
 bit                      4.0.5     2022-11-15 [1] CRAN (R 4.3.0)
 bit64                    4.0.5     2020-08-30 [1] CRAN (R 4.3.0)
 bitops                   1.0-7     2021-04-24 [1] CRAN (R 4.3.0)
 blob                     1.2.3     2022-04-10 [1] CRAN (R 4.3.0)
 cachem                   1.0.7     2023-02-24 [1] CRAN (R 4.3.0)
 callr                    3.7.3     2022-11-02 [1] CRAN (R 4.3.0)
 cli                      3.6.0     2023-01-09 [1] CRAN (R 4.3.0)
 colorspace               2.1-0     2023-01-23 [1] CRAN (R 4.3.0)
 cowplot                * 1.1.1     2020-12-30 [1] CRAN (R 4.3.0)
 crayon                   1.5.2     2022-09-29 [1] CRAN (R 4.3.0)
 curl                     5.0.0     2023-01-12 [1] CRAN (R 4.3.0)
 DBI                      1.1.3     2022-06-18 [1] CRAN (R 4.3.0)
 dbplyr                 * 2.3.1     2023-02-24 [1] CRAN (R 4.3.0)
 devtools                 2.4.5     2022-10-11 [1] CRAN (R 4.3.0)
 digest                   0.6.31    2022-12-11 [1] CRAN (R 4.3.0)
 dplyr                  * 1.1.0     2023-01-29 [1] CRAN (R 4.3.0)
 ellipsis                 0.3.2     2021-04-29 [1] CRAN (R 4.3.0)
 evaluate                 0.20      2023-01-17 [1] CRAN (R 4.3.0)
 fansi                    1.0.4     2023-01-22 [1] CRAN (R 4.3.0)
 fastmap                  1.1.1     2023-02-24 [1] CRAN (R 4.3.0)
 filelock                 1.0.2     2018-10-05 [1] CRAN (R 4.3.0)
 forcats                * 1.0.0     2023-01-29 [1] CRAN (R 4.3.0)
 fs                       1.6.1     2023-02-06 [1] CRAN (R 4.3.0)
 generics                 0.1.3     2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb             1.35.15   2023-02-03 [1] Bioconductor
 GenomeInfoDbData         1.2.9     2022-11-04 [1] Bioconductor
 ggplot2                * 3.4.1     2023-02-10 [1] CRAN (R 4.3.0)
 glue                     1.6.2     2022-02-24 [1] CRAN (R 4.3.0)
 gtable                   0.3.1     2022-09-01 [1] CRAN (R 4.3.0)
 hms                      1.1.2     2022-08-19 [1] CRAN (R 4.3.0)
 htmltools                0.5.4     2022-12-07 [1] CRAN (R 4.3.0)
 htmlwidgets              1.6.1     2023-01-07 [1] CRAN (R 4.3.0)
 httpuv                   1.6.9     2023-02-14 [1] CRAN (R 4.3.0)
 httr                     1.4.5     2023-02-24 [1] CRAN (R 4.3.0)
 interactiveDisplayBase   1.37.0    2022-11-01 [1] Bioconductor
 IRanges                * 2.33.0    2022-11-01 [1] Bioconductor
 jsonlite                 1.8.4     2022-12-06 [1] CRAN (R 4.3.0)
 KEGGREST                 1.39.0    2022-11-01 [1] Bioconductor
 knitr                    1.42      2023-01-25 [1] CRAN (R 4.3.0)
 later                    1.3.0     2021-08-18 [1] CRAN (R 4.3.0)
 lifecycle                1.0.3     2022-10-07 [1] CRAN (R 4.3.0)
 lubridate              * 1.9.2     2023-02-10 [1] CRAN (R 4.3.0)
 magrittr                 2.0.3     2022-03-30 [1] CRAN (R 4.3.0)
 memoise                  2.0.1     2021-11-26 [1] CRAN (R 4.3.0)
 mime                     0.12      2021-09-28 [1] CRAN (R 4.3.0)
 miniUI                   0.1.1.1   2018-05-18 [1] CRAN (R 4.3.0)
 munsell                  0.5.0     2018-06-12 [1] CRAN (R 4.3.0)
 pillar                   1.8.1     2022-08-19 [1] CRAN (R 4.3.0)
 pkgbuild                 1.4.0     2022-11-27 [1] CRAN (R 4.3.0)
 pkgconfig                2.0.3     2019-09-22 [1] CRAN (R 4.3.0)
 pkgload                  1.3.2     2022-11-16 [1] CRAN (R 4.3.0)
 png                      0.1-8     2022-11-29 [1] CRAN (R 4.3.0)
 prettyunits              1.1.1     2020-01-24 [1] CRAN (R 4.3.0)
 processx                 3.8.0     2022-10-26 [1] CRAN (R 4.3.0)
 profvis                  0.3.7     2020-11-02 [1] CRAN (R 4.3.0)
 promises                 1.2.0.1   2021-02-11 [1] CRAN (R 4.3.0)
 ps                       1.7.2     2022-10-26 [1] CRAN (R 4.3.0)
 purrr                  * 1.0.1     2023-01-10 [1] CRAN (R 4.3.0)
 R6                       2.5.1     2021-08-19 [1] CRAN (R 4.3.0)
 rappdirs                 0.3.3     2021-01-31 [1] CRAN (R 4.3.0)
 Rcpp                     1.0.10    2023-01-22 [1] CRAN (R 4.3.0)
 RCurl                    1.98-1.10 2023-01-27 [1] CRAN (R 4.3.0)
 readr                  * 2.1.4     2023-02-10 [1] CRAN (R 4.3.0)
 remotes                  2.4.2     2021-11-30 [1] CRAN (R 4.3.0)
 rlang                    1.0.6     2022-09-24 [1] CRAN (R 4.3.0)
 rmarkdown                2.20      2023-01-19 [1] CRAN (R 4.3.0)
 RSQLite                  2.3.0     2023-02-17 [1] CRAN (R 4.3.0)
 S4Vectors              * 0.37.4    2023-02-26 [1] Bioconductor
 scales                   1.2.1     2022-08-20 [1] CRAN (R 4.3.0)
 sessioninfo              1.2.2     2021-12-06 [1] CRAN (R 4.3.0)
 shiny                    1.7.4     2022-12-15 [1] CRAN (R 4.3.0)
 stringi                  1.7.12    2023-01-11 [1] CRAN (R 4.3.0)
 stringr                * 1.5.0     2022-12-02 [1] CRAN (R 4.3.0)
 tibble                 * 3.1.8     2022-07-22 [1] CRAN (R 4.3.0)
 tidyr                  * 1.3.0     2023-01-24 [1] CRAN (R 4.3.0)
 tidyselect               1.2.0     2022-10-10 [1] CRAN (R 4.3.0)
 tidyverse              * 2.0.0     2023-02-22 [1] CRAN (R 4.3.0)
 timechange               0.2.0     2023-01-11 [1] CRAN (R 4.3.0)
 tzdb                     0.3.0     2022-03-28 [1] CRAN (R 4.3.0)
 urlchecker               1.0.1     2021-11-30 [1] CRAN (R 4.3.0)
 usethis                  2.1.6     2022-05-25 [1] CRAN (R 4.3.0)
 utf8                     1.2.3     2023-01-31 [1] CRAN (R 4.3.0)
 vctrs                    0.5.2     2023-01-23 [1] CRAN (R 4.3.0)
 withr                    2.5.0     2022-03-03 [1] CRAN (R 4.3.0)
 xfun                     0.37      2023-01-31 [1] CRAN (R 4.3.0)
 xtable                   1.8-4     2019-04-21 [1] CRAN (R 4.3.0)
 XVector                  0.39.0    2022-12-20 [1] Bioconductor
 yaml                     2.3.7     2023-01-23 [1] CRAN (R 4.3.0)
 zlibbioc                 1.45.0    2022-12-20 [1] Bioconductor

 [1] /Users/jacques/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../content/day3/Lecture4_clustering.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Lecture 4 - Identifying cell populations</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../content/day3/Lecture5_batchcorrection.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Lecture 5 - Data integration and batch effect correction</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">Single-cell RNAseq analysis with R/Bioconductor |<br>
J. Serizay O. Ashenberg</div>   
    <div class="nav-footer-right">This book was built with <a href="https://quarto.org/">Quarto</a>.</div>
  </div>
</footer>


</body></html>