<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.189">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="J. Serizay &amp; O. Ashenberg">

<title>Single-cell RNAseq analysis with R/Bioconductor - 13&nbsp; Lab 7: Pseudotime analyses</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../content/day4/Lecture7_ATAC.html" rel="next">
<link href="../../content/day4/Lecture6_pseudotime.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Lab 7: Pseudotime analyses</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">Single-cell RNAseq analysis with R/Bioconductor</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/js2264/scRNAseq_Physalia_2023/" title="Source Code" class="sidebar-tool px-1"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">Welcome</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/program.html" class="sidebar-item-text sidebar-link">Program</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/Rstudio.html" class="sidebar-item-text sidebar-link">RStudio</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/prerequisites.html" class="sidebar-item-text sidebar-link">Prerequisites</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Day 1</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lecture1_introduction.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Lecture 1 - Introduction to scRNAseq analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lab1_Intro.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Lab 1: Familiarizing yourself with the course AWS instance</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lecture2_Bcl2matrix.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Lecture 2 - From sequencing reads to expression matrices</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lab2_processingreads.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Lab 2: From .bcl to count matrix</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">Day 2</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lab3_Rbioc.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Lab 3: Introduction to R/Bioconductor</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lecture3_qualitycontrol.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Lecture 3 - Quality control for scRNA-Seq data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lab4_data_wrangling_scRNAseq.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Lab 4 - Single-cell RNA-seq data wrangling</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">Day 3</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lecture4_clustering.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Lecture 4 - Identifying cell populations</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lab5_clustering.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Lab 5: Dimension reduction, clustering and annotation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lecture5_batchcorrection.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Lecture 5 - Data integration and batch effect correction</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lab6_batch_correction.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Lab 6 - Batch correction</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">Day 4</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lecture6_pseudotime.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Lecture 6 - Trajectories and pseudotimes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lab7_pseudotime.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Lab 7: Pseudotime analyses</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lecture7_ATAC.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Lecture 7 - Advances in single-cell genomics: the epigenome</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lab8_atac-seq.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Lab 8 - Single-cell ATAC-seq analysis workflow</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">Day 5</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day5/Lecture8_spatial-transcriptomics.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Lecture 8 - Advances in single-cell genomics: spatial transcriptomics</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/extra.html" class="sidebar-item-text sidebar-link">Extra resources</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#process-testis-data-in-r" id="toc-process-testis-data-in-r" class="nav-link active" data-scroll-target="#process-testis-data-in-r"><span class="toc-section-number">13.1</span>  Process testis data in R</a>
  <ul class="collapse">
  <li><a href="#import-testis-data-from-gse112013-and-pre-process-it" id="toc-import-testis-data-from-gse112013-and-pre-process-it" class="nav-link" data-scroll-target="#import-testis-data-from-gse112013-and-pre-process-it"><span class="toc-section-number">13.1.1</span>  Import testis data from GSE112013 and pre-process it</a></li>
  <li><a href="#annotate-cells-using-hpa-resources" id="toc-annotate-cells-using-hpa-resources" class="nav-link" data-scroll-target="#annotate-cells-using-hpa-resources"><span class="toc-section-number">13.1.2</span>  Annotate cells using HPA resources</a></li>
  <li><a href="#filter-the-testis-dataset-to-only-germinal-cells." id="toc-filter-the-testis-dataset-to-only-germinal-cells." class="nav-link" data-scroll-target="#filter-the-testis-dataset-to-only-germinal-cells."><span class="toc-section-number">13.1.3</span>  Filter the testis dataset to only germinal cells.</a></li>
  </ul></li>
  <li><a href="#trajectory-inference-ti-in-scrnaseq" id="toc-trajectory-inference-ti-in-scrnaseq" class="nav-link" data-scroll-target="#trajectory-inference-ti-in-scrnaseq"><span class="toc-section-number">13.2</span>  Trajectory inference (TI) in scRNAseq</a>
  <ul class="collapse">
  <li><a href="#trajectory" id="toc-trajectory" class="nav-link" data-scroll-target="#trajectory"><span class="toc-section-number">13.2.1</span>  Trajectory</a></li>
  <li><a href="#pseudotime" id="toc-pseudotime" class="nav-link" data-scroll-target="#pseudotime"><span class="toc-section-number">13.2.2</span>  Pseudotime</a></li>
  <li><a href="#bonus-daunting-snippet-but-that-makes-a-cool-figure-for-a-paper-modeling-pseudotime-dependent-gene-expression" id="toc-bonus-daunting-snippet-but-that-makes-a-cool-figure-for-a-paper-modeling-pseudotime-dependent-gene-expression" class="nav-link" data-scroll-target="#bonus-daunting-snippet-but-that-makes-a-cool-figure-for-a-paper-modeling-pseudotime-dependent-gene-expression"><span class="toc-section-number">13.2.3</span>  BONUS: Daunting snippet but that makes a cool figure for a paper: modeling pseudotime-dependent gene expression</a></li>
  </ul></li>
  <li><a href="#ordering-trajectory-with-rna-velocity" id="toc-ordering-trajectory-with-rna-velocity" class="nav-link" data-scroll-target="#ordering-trajectory-with-rna-velocity"><span class="toc-section-number">13.3</span>  Ordering trajectory with RNA velocity</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info"><span class="toc-section-number">13.4</span>  Session info</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/js2264/scRNAseq_Physalia_2023/edit/main/content/day4/Lab7_pseudotime.qmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/js2264/scRNAseq_Physalia_2023/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Lab 7: Pseudotime analyses</span></h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<div class="cell">

</div>
<div class="cell">

</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Aims
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Understand the requirements for RNA velocity computation</li>
<li>Process scRNAseq using ‘spliced’ counts</li>
<li>Perform lineage inference</li>
<li>Compute RNA velocity and use it to orientate lineages</li>
</ul>
</div>
</div>
<section id="process-testis-data-in-r" class="level2" data-number="13.1">
<h2 data-number="13.1" class="anchored" data-anchor-id="process-testis-data-in-r"><span class="header-section-number">13.1</span> Process testis data in R</h2>
<section id="import-testis-data-from-gse112013-and-pre-process-it" class="level3" data-number="13.1.1">
<h3 data-number="13.1.1" class="anchored" data-anchor-id="import-testis-data-from-gse112013-and-pre-process-it"><span class="header-section-number">13.1.1</span> Import testis data from GSE112013 and pre-process it</h3>
<p>The workflow we have covered in the previous days can be reused here.</p>
<ul>
<li><strong>Import testis dataset in R, remove doublets, filter genes, normalize <code>counts</code>, correct for batch effect and cluster cells.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SingleCellExperiment)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Download raw file</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">'Guo_testis'</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">'https://ftp.ncbi.nlm.nih.gov/geo/series/GSE112nnn/GSE112013/suppl/GSE112013_Combined_UMI_table.txt.gz'</span>, </span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">'Guo_testis/GSE112013_Combined_UMI_table.txt.gz'</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(<span class="st">'gunzip Guo_testis/GSE112013_Combined_UMI_table.txt.gz'</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Create SingleCellExperiment object</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> vroom<span class="sc">::</span><span class="fu">vroom</span>(<span class="st">'Guo_testis/GSE112013_Combined_UMI_table.txt'</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(x[, <span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">as</span>(counts, <span class="st">'dgCMatrix'</span>)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(x[, <span class="dv">1</span>])</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>cells <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">cellid =</span> <span class="fu">colnames</span>(x[, <span class="sc">-</span><span class="dv">1</span>]))</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>testis <span class="ot">&lt;-</span> <span class="fu">SingleCellExperiment</span>(</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> counts), </span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">colData =</span> cells, </span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">rowData =</span> genes</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>testis<span class="sc">$</span>Barcode <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(testis<span class="sc">$</span>cellid, <span class="st">'Donor.-'</span>, <span class="st">''</span>) <span class="sc">|&gt;</span> <span class="fu">str_replace</span>(<span class="st">'-.'</span>, <span class="st">''</span>)</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>testis <span class="ot">&lt;-</span> testis[, <span class="sc">!</span><span class="fu">duplicated</span>(testis<span class="sc">$</span>Barcode)]</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>testis<span class="sc">$</span>donor <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(testis<span class="sc">$</span>cellid, <span class="st">'-.*'</span>, <span class="st">''</span>)</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>testis<span class="sc">$</span>replicate <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(testis<span class="sc">$</span>cellid, <span class="st">'.*-'</span>, <span class="st">''</span>)</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(testis) <span class="ot">&lt;-</span> <span class="fu">rowData</span>(testis)<span class="sc">$</span>Gene</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1000</span>)</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove doublets</span></span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>testis <span class="ot">&lt;-</span> scDblFinder<span class="sc">::</span><span class="fu">scDblFinder</span>(testis)</span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a>testis <span class="ot">&lt;-</span> testis[, testis<span class="sc">$</span>scDblFinder.class <span class="sc">==</span> <span class="st">'singlet'</span>]</span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Recover human genomic, protein-coding gene informations</span></span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plyranges)</span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a>ah <span class="ot">&lt;-</span> AnnotationHub<span class="sc">::</span><span class="fu">AnnotationHub</span>()</span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a>AnnotationHub<span class="sc">::</span><span class="fu">query</span>(ah, <span class="fu">c</span>(<span class="st">'gene annotation'</span>, <span class="st">'ensembl'</span>, <span class="st">'102'</span>, <span class="st">'homo_sapiens'</span>, <span class="st">'GRCh38'</span>))</span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a>gtf <span class="ot">&lt;-</span> AnnotationHub<span class="sc">::</span><span class="fu">query</span>(ah, <span class="fu">c</span>(<span class="st">'Homo_sapiens.GRCh38.102.chr.gtf'</span>))[[<span class="dv">1</span>]]</span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> gtf <span class="sc">|&gt;</span> </span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(type <span class="sc">==</span> <span class="st">'gene'</span>) <span class="sc">|&gt;</span> </span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(gene_biotype <span class="sc">==</span> <span class="st">'protein_coding'</span>) <span class="sc">|&gt;</span> </span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(gene_source <span class="sc">==</span> <span class="st">'ensembl_havana'</span>) </span>
<span id="cb1-43"><a href="#cb1-43" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> genes[<span class="sc">!</span><span class="fu">duplicated</span>(genes<span class="sc">$</span>gene_name)]</span>
<span id="cb1-44"><a href="#cb1-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-45"><a href="#cb1-45" aria-hidden="true" tabindex="-1"></a><span class="co"># Annotate genes in testis dataset and filter out non-relevant genes</span></span>
<span id="cb1-46"><a href="#cb1-46" aria-hidden="true" tabindex="-1"></a>testis <span class="ot">&lt;-</span> testis[genes<span class="sc">$</span>gene_name[genes<span class="sc">$</span>gene_name <span class="sc">%in%</span> <span class="fu">rownames</span>(testis)], ]</span>
<span id="cb1-47"><a href="#cb1-47" aria-hidden="true" tabindex="-1"></a><span class="fu">rowRanges</span>(testis) <span class="ot">&lt;-</span> genes[<span class="fu">match</span>(<span class="fu">rownames</span>(testis), genes<span class="sc">$</span>gene_name)]</span>
<span id="cb1-48"><a href="#cb1-48" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(testis) <span class="ot">&lt;-</span> <span class="fu">rowData</span>(testis)[, <span class="fu">c</span>(<span class="st">'gene_name'</span>, <span class="st">'gene_id'</span>)]</span>
<span id="cb1-49"><a href="#cb1-49" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(testis) <span class="ot">&lt;-</span> scuttle<span class="sc">::</span><span class="fu">uniquifyFeatureNames</span>(<span class="fu">rowData</span>(testis)<span class="sc">$</span>gene_id, <span class="fu">rowData</span>(testis)<span class="sc">$</span>gene_name)</span>
<span id="cb1-50"><a href="#cb1-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-51"><a href="#cb1-51" aria-hidden="true" tabindex="-1"></a><span class="co"># Get preliminary QCs per cell and per gene</span></span>
<span id="cb1-52"><a href="#cb1-52" aria-hidden="true" tabindex="-1"></a>testis <span class="ot">&lt;-</span> scuttle<span class="sc">::</span><span class="fu">addPerCellQCMetrics</span>(testis)</span>
<span id="cb1-53"><a href="#cb1-53" aria-hidden="true" tabindex="-1"></a>testis <span class="ot">&lt;-</span> scuttle<span class="sc">::</span><span class="fu">addPerFeatureQCMetrics</span>(testis)</span>
<span id="cb1-54"><a href="#cb1-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-55"><a href="#cb1-55" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter out genes not expressed in at least 10 cells</span></span>
<span id="cb1-56"><a href="#cb1-56" aria-hidden="true" tabindex="-1"></a>testis <span class="ot">&lt;-</span> testis[<span class="fu">rowSums</span>(<span class="fu">assay</span>(testis, <span class="st">'counts'</span>) <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&gt;=</span> <span class="dv">10</span>, ]</span>
<span id="cb1-57"><a href="#cb1-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-58"><a href="#cb1-58" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize counts using VST</span></span>
<span id="cb1-59"><a href="#cb1-59" aria-hidden="true" tabindex="-1"></a>cnts <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">assay</span>(testis, <span class="st">'counts'</span>), <span class="st">'dgCMatrix'</span>)</span>
<span id="cb1-60"><a href="#cb1-60" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(cnts) <span class="ot">&lt;-</span> testis<span class="sc">$</span>cellid</span>
<span id="cb1-61"><a href="#cb1-61" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(cnts) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(testis)</span>
<span id="cb1-62"><a href="#cb1-62" aria-hidden="true" tabindex="-1"></a>testis_vst <span class="ot">&lt;-</span> sctransform<span class="sc">::</span><span class="fu">vst</span>(cnts, <span class="at">return_cell_attr =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-63"><a href="#cb1-63" aria-hidden="true" tabindex="-1"></a>corrected_cnts <span class="ot">&lt;-</span> sctransform<span class="sc">::</span><span class="fu">correct</span>(testis_vst)</span>
<span id="cb1-64"><a href="#cb1-64" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(testis, <span class="st">'corrected_counts'</span>, <span class="at">withDimnames =</span> <span class="cn">FALSE</span>) <span class="ot">&lt;-</span> corrected_cnts</span>
<span id="cb1-65"><a href="#cb1-65" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(testis, <span class="st">'logcounts'</span>, <span class="at">withDimnames =</span> <span class="cn">FALSE</span>) <span class="ot">&lt;-</span> <span class="fu">log1p</span>(corrected_cnts)</span>
<span id="cb1-66"><a href="#cb1-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-67"><a href="#cb1-67" aria-hidden="true" tabindex="-1"></a><span class="co"># Computing biological variance of each gene</span></span>
<span id="cb1-68"><a href="#cb1-68" aria-hidden="true" tabindex="-1"></a>testis_variance <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">modelGeneVar</span>(testis)</span>
<span id="cb1-69"><a href="#cb1-69" aria-hidden="true" tabindex="-1"></a>HVGs <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">getTopHVGs</span>(testis_variance, <span class="at">prop =</span> <span class="fl">0.1</span>)</span>
<span id="cb1-70"><a href="#cb1-70" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(testis)<span class="sc">$</span>isHVG <span class="ot">&lt;-</span> <span class="fu">rownames</span>(testis) <span class="sc">%in%</span> HVGs</span>
<span id="cb1-71"><a href="#cb1-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-72"><a href="#cb1-72" aria-hidden="true" tabindex="-1"></a><span class="co"># Embedding dataset in PCA space, correcting for batch effect</span></span>
<span id="cb1-73"><a href="#cb1-73" aria-hidden="true" tabindex="-1"></a>mergedBatches <span class="ot">&lt;-</span> batchelor<span class="sc">::</span><span class="fu">fastMNN</span>(</span>
<span id="cb1-74"><a href="#cb1-74" aria-hidden="true" tabindex="-1"></a>    testis, </span>
<span id="cb1-75"><a href="#cb1-75" aria-hidden="true" tabindex="-1"></a>    <span class="at">batch =</span> testis<span class="sc">$</span>donor, </span>
<span id="cb1-76"><a href="#cb1-76" aria-hidden="true" tabindex="-1"></a>    <span class="at">subset.row =</span> HVGs, </span>
<span id="cb1-77"><a href="#cb1-77" aria-hidden="true" tabindex="-1"></a>    <span class="at">BPPARAM =</span> BiocParallel<span class="sc">::</span><span class="fu">MulticoreParam</span>(<span class="at">workers =</span> <span class="dv">12</span>)</span>
<span id="cb1-78"><a href="#cb1-78" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-79"><a href="#cb1-79" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDim</span>(testis, <span class="st">'corrected'</span>) <span class="ot">&lt;-</span> <span class="fu">reducedDim</span>(mergedBatches, <span class="st">'corrected'</span>)</span>
<span id="cb1-80"><a href="#cb1-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-81"><a href="#cb1-81" aria-hidden="true" tabindex="-1"></a><span class="co"># Embedding dataset in shared k-nearest neighbors graph for clustering </span></span>
<span id="cb1-82"><a href="#cb1-82" aria-hidden="true" tabindex="-1"></a>graph <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">buildSNNGraph</span>(testis, <span class="at">use.dimred =</span> <span class="st">'corrected'</span>)</span>
<span id="cb1-83"><a href="#cb1-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-84"><a href="#cb1-84" aria-hidden="true" tabindex="-1"></a><span class="co"># Cluster cells using Louvain community finding algorithm</span></span>
<span id="cb1-85"><a href="#cb1-85" aria-hidden="true" tabindex="-1"></a>testis_clust <span class="ot">&lt;-</span> igraph<span class="sc">::</span><span class="fu">cluster_louvain</span>(graph)<span class="sc">$</span>membership</span>
<span id="cb1-86"><a href="#cb1-86" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(testis_clust)</span>
<span id="cb1-87"><a href="#cb1-87" aria-hidden="true" tabindex="-1"></a>testis<span class="sc">$</span>cluster <span class="ot">&lt;-</span> <span class="fu">factor</span>(testis_clust)</span>
<span id="cb1-88"><a href="#cb1-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-89"><a href="#cb1-89" aria-hidden="true" tabindex="-1"></a><span class="co"># Embedding dataset in TSNE space for visualization</span></span>
<span id="cb1-90"><a href="#cb1-90" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">10</span>)</span>
<span id="cb1-91"><a href="#cb1-91" aria-hidden="true" tabindex="-1"></a>testis <span class="ot">&lt;-</span> scater<span class="sc">::</span><span class="fu">runTSNE</span>(testis, <span class="at">dimred =</span> <span class="st">'corrected'</span>)</span>
<span id="cb1-92"><a href="#cb1-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-93"><a href="#cb1-93" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize embeddings</span></span>
<span id="cb1-94"><a href="#cb1-94" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(</span>
<span id="cb1-95"><a href="#cb1-95" aria-hidden="true" tabindex="-1"></a>    scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(testis, <span class="st">'corrected'</span>, <span class="at">colour_by =</span> <span class="st">'donor'</span>),</span>
<span id="cb1-96"><a href="#cb1-96" aria-hidden="true" tabindex="-1"></a>    scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(testis, <span class="st">'corrected'</span>, <span class="at">colour_by =</span> <span class="st">'cluster'</span>),</span>
<span id="cb1-97"><a href="#cb1-97" aria-hidden="true" tabindex="-1"></a>    scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(testis, <span class="st">'TSNE'</span>, <span class="at">colour_by =</span> <span class="st">'donor'</span>),</span>
<span id="cb1-98"><a href="#cb1-98" aria-hidden="true" tabindex="-1"></a>    scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(testis, <span class="st">'TSNE'</span>, <span class="at">colour_by =</span> <span class="st">'cluster'</span>)</span>
<span id="cb1-99"><a href="#cb1-99" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="annotate-cells-using-hpa-resources" class="level3" data-number="13.1.2">
<h3 data-number="13.1.2" class="anchored" data-anchor-id="annotate-cells-using-hpa-resources"><span class="header-section-number">13.1.2</span> Annotate cells using HPA resources</h3>
<ul>
<li><strong>Load <code>HPA</code> data <a href="https://www.proteinatlas.org/download/rna_single_cell_type.tsv.zip">from internet</a>.</strong></li>
<li><strong>Try to format it as a <code>SummarizedExperiment</code>.</strong></li>
<li><strong>What celltypes are profiled?</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">'https://www.proteinatlas.org/download/rna_single_cell_type.tsv.zip'</span>, </span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">'Guo_testis/rna_single_cell_type.tsv.zip'</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(<span class="st">'unzip Guo_testis/rna_single_cell_type.tsv.zip'</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(<span class="st">'mv rna_single_cell_type.tsv Guo_testis/'</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>HPA <span class="ot">&lt;-</span> vroom<span class="sc">::</span><span class="fu">vroom</span>(<span class="st">'Guo_testis/rna_single_cell_type.tsv'</span>) <span class="sc">|&gt;</span> </span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> <span class="st">`</span><span class="at">Cell type</span><span class="st">`</span>, <span class="at">values_from =</span> <span class="st">'nTPM'</span>) <span class="sc">|&gt;</span> </span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>Gene) <span class="sc">|&gt;</span> </span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">distinct</span>(<span class="st">`</span><span class="at">Gene name</span><span class="st">`</span>, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span> </span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">column_to_rownames</span>(<span class="st">'Gene name'</span>)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>HPA_se <span class="ot">&lt;-</span> SummarizedExperiment<span class="sc">::</span><span class="fu">SummarizedExperiment</span>(HPA, <span class="at">colData =</span> <span class="fu">tibble</span>(<span class="at">cell_type =</span> <span class="fu">colnames</span>(HPA)), <span class="at">assays =</span> <span class="fu">list</span>(<span class="st">'logcounts'</span> <span class="ot">=</span> HPA))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>Use these cell type profiles to annotate cell types in the testis dataset.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Transfer annotations to `testis`</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>predictions <span class="ot">&lt;-</span> SingleR<span class="sc">::</span><span class="fu">SingleR</span>(</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">test =</span> testis, </span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">ref =</span> HPA_se, </span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels =</span> <span class="fu">colData</span>(HPA_se)<span class="sc">$</span>cell_type, </span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">clusters =</span> testis<span class="sc">$</span>cluster</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>testis<span class="sc">$</span>annotation <span class="ot">&lt;-</span> predictions<span class="sc">$</span>labels[testis<span class="sc">$</span>cluster]</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>    scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(testis, <span class="at">dimred =</span> <span class="st">'corrected'</span>, <span class="at">colour_by =</span> <span class="st">'cluster'</span>, <span class="at">text_by =</span> <span class="st">'cluster'</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">'Testis data, PCA, graph-based clusters'</span>), </span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>    scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(testis, <span class="at">dimred =</span> <span class="st">'corrected'</span>, <span class="at">colour_by =</span> <span class="st">'annotation'</span>, <span class="at">text_by =</span> <span class="st">'annotation'</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">'PCA, Annotations transferred from HPA'</span>),</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(testis, <span class="at">dimred =</span> <span class="st">'TSNE'</span>, <span class="at">colour_by =</span> <span class="st">'cluster'</span>, <span class="at">text_by =</span> <span class="st">'cluster'</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">'Testis data, tSNE, graph-based clusters'</span>), </span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>    scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(testis, <span class="at">dimred =</span> <span class="st">'TSNE'</span>, <span class="at">colour_by =</span> <span class="st">'annotation'</span>, <span class="at">text_by =</span> <span class="st">'annotation'</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">'tSNE, Annotations transferred from HPA'</span>)</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="filter-the-testis-dataset-to-only-germinal-cells." class="level3" data-number="13.1.3">
<h3 data-number="13.1.3" class="anchored" data-anchor-id="filter-the-testis-dataset-to-only-germinal-cells."><span class="header-section-number">13.1.3</span> Filter the testis dataset to only germinal cells.</h3>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>germcells <span class="ot">&lt;-</span> testis[</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    , </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    testis<span class="sc">$</span>annotation <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Spermatogonia"</span>, <span class="st">"Spermatocytes"</span>, <span class="st">"Early spermatids"</span>, <span class="st">"Late spermatids"</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
</section>
<section id="trajectory-inference-ti-in-scrnaseq" class="level2" data-number="13.2">
<h2 data-number="13.2" class="anchored" data-anchor-id="trajectory-inference-ti-in-scrnaseq"><span class="header-section-number">13.2</span> Trajectory inference (TI) in scRNAseq</h2>
<p>An important question in scRNAseq field of research is: how to identify a cell trajectory from high-dimensional expression data and map individual cells onto it? A large number of methods have currently emerged, each one with their own specificities, assumptions, and strengths. A nice breakdown (from 2019, so already very outdated!) is available from <code>Saelens et al., Nat. Biotech. 2018</code> (doi: 10.1038/s41587-019-0071-9):</p>
<p><img src="https://github.com/js2264/scRNAseq_Physalia_2023/blob/main/content/day4/.img/2021-06-01-11-29-36.png?raw=true"></p>
<section id="trajectory" class="level3" data-number="13.2.1">
<h3 data-number="13.2.1" class="anchored" data-anchor-id="trajectory"><span class="header-section-number">13.2.1</span> Trajectory</h3>
<p><code>Slingshot</code> is perhaps one of the most widely used algorithms for users who want to focus on <code>R</code>-based approaches.</p>
<ul>
<li><strong>Read Slingshot documentation to understand how to identify lineages in a scRNAseq dataset in R</strong></li>
<li><strong>Why is it recommended to infer lineages from PCA space rather than t-SNE or UMAP space, even though these spaces do “reveal” an obvious trajectory in 2D?</strong></li>
<li><strong>Infer lineages, using cluster annotations as information to build the MST. Note that you will first need to remove the 50th PCA dimension for <code>slingshot</code> to work (bug reported).</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDim</span>(germcells, <span class="st">'corrected_2'</span>) <span class="ot">&lt;-</span> <span class="fu">reducedDim</span>(germcells, <span class="st">'corrected'</span>)[, <span class="dv">1</span><span class="sc">:</span><span class="dv">49</span>]</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>germcells_slingshot <span class="ot">&lt;-</span> slingshot<span class="sc">::</span><span class="fu">slingshot</span>(</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    germcells, </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">reducedDim =</span> <span class="st">'corrected_2'</span>, </span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">clusterLabels =</span> germcells<span class="sc">$</span>annotation</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>slingshot<span class="sc">::</span><span class="fu">slingLineages</span>(germcells_slingshot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="pseudotime" class="level3" data-number="13.2.2">
<h3 data-number="13.2.2" class="anchored" data-anchor-id="pseudotime"><span class="header-section-number">13.2.2</span> Pseudotime</h3>
<ul>
<li><strong>Recover pseudotime values and principal curves from <code>slingshot</code> output</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>germcells<span class="sc">$</span>pseudotime <span class="ot">&lt;-</span> slingshot<span class="sc">::</span><span class="fu">slingPseudotime</span>(germcells_slingshot)[, <span class="st">'Lineage1'</span>]</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>pca_curve <span class="ot">&lt;-</span> slingshot<span class="sc">::</span><span class="fu">slingCurves</span>(germcells_slingshot, <span class="at">as.df =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(pca_curve) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">'PC'</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(pca_curve))</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>tsne_curve <span class="ot">&lt;-</span> slingshot<span class="sc">::</span><span class="fu">embedCurves</span>(germcells_slingshot, <span class="st">'TSNE'</span>, <span class="at">smoother =</span> <span class="st">'loess'</span>, <span class="at">span =</span> <span class="fl">0.1</span>) <span class="sc">|&gt;</span> slingshot<span class="sc">::</span><span class="fu">slingCurves</span>(<span class="at">as.df =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>tsne_curve <span class="ot">&lt;-</span> tsne_curve[<span class="fu">order</span>(tsne_curve<span class="sc">$</span>Order), ]</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tsne_curve) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">'TSNE'</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(tsne_curve))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>Plot PCA and tSNE embeddings, coloring cells by either their annotation or their pseudotime value. Overlay the principal curves to each embedding</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">PC1 =</span> <span class="fu">reducedDim</span>(germcells, <span class="st">'corrected'</span>)[,<span class="dv">1</span>], </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">PC2 =</span> <span class="fu">reducedDim</span>(germcells, <span class="st">'corrected'</span>)[,<span class="dv">2</span>], </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">TSNE1 =</span> <span class="fu">reducedDim</span>(germcells, <span class="st">'TSNE'</span>)[,<span class="dv">1</span>], </span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">TSNE2 =</span> <span class="fu">reducedDim</span>(germcells, <span class="st">'TSNE'</span>)[,<span class="dv">2</span>], </span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotation =</span> germcells<span class="sc">$</span>annotation, </span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">pseudotime =</span> germcells<span class="sc">$</span>pseudotime</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>    df <span class="sc">|&gt;</span> </span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_point</span>(<span class="fu">aes</span>(PC1, PC2, <span class="at">col =</span> annotation)) <span class="sc">+</span> </span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_path</span>(<span class="at">data =</span> pca_curve, <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2)) <span class="sc">+</span> </span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>        <span class="fu">coord_fixed</span>(),</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>    df <span class="sc">|&gt;</span> </span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_point</span>(<span class="fu">aes</span>(TSNE1, TSNE2, <span class="at">col =</span> annotation)) <span class="sc">+</span> </span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_path</span>(<span class="at">data =</span> tsne_curve, <span class="fu">aes</span>(<span class="at">x =</span> TSNE1, <span class="at">y =</span> TSNE2)) <span class="sc">+</span> </span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>        <span class="fu">coord_fixed</span>(),</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>    df <span class="sc">|&gt;</span> </span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_point</span>(<span class="fu">aes</span>(PC1, PC2, <span class="at">col =</span> pseudotime)) <span class="sc">+</span> </span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_path</span>(<span class="at">data =</span> pca_curve, <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2)) <span class="sc">+</span> </span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>        <span class="fu">coord_fixed</span>(),</span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>    df <span class="sc">|&gt;</span> </span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_point</span>(<span class="fu">aes</span>(TSNE1, TSNE2, <span class="at">col =</span> pseudotime)) <span class="sc">+</span> </span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_path</span>(<span class="at">data =</span> tsne_curve, <span class="fu">aes</span>(<span class="at">x =</span> TSNE1, <span class="at">y =</span> TSNE2)) <span class="sc">+</span> </span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a>        <span class="fu">coord_fixed</span>()</span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>Check the distribution of pseudotime values across the different cell clusters.</strong></li>
<li><strong>What do you observe? Where you expecting this?</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotation =</span> <span class="fu">factor</span>(germcells<span class="sc">$</span>annotation, <span class="fu">c</span>(<span class="st">"Spermatogonia"</span>, <span class="st">"Spermatocytes"</span>, <span class="st">"Early spermatids"</span>, <span class="st">"Late spermatids"</span>)), </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">pseudotime =</span> germcells<span class="sc">$</span>pseudotime</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span> </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> annotation, <span class="at">y =</span> pseudotime, <span class="at">fill =</span> annotation)) <span class="sc">+</span> </span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_violin</span>(<span class="at">scale =</span> <span class="st">'width'</span>) <span class="sc">+</span> </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="at">outlier.shape =</span> <span class="cn">NULL</span>, <span class="at">width =</span> <span class="fl">0.1</span>, <span class="at">fill =</span> <span class="st">'white'</span>) <span class="sc">+</span> </span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">vjust =</span> <span class="dv">1</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>Correct pseudotime values as you would expect it to be.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>germcells<span class="sc">$</span>pseudotime <span class="ot">&lt;-</span> scales<span class="sc">::</span><span class="fu">rescale</span>((<span class="sc">-</span><span class="dv">1</span> <span class="sc">*</span> slingshot<span class="sc">::</span><span class="fu">slingPseudotime</span>(germcells_slingshot)[, <span class="st">'Lineage1'</span>]), <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="bonus-daunting-snippet-but-that-makes-a-cool-figure-for-a-paper-modeling-pseudotime-dependent-gene-expression" class="level3" data-number="13.2.3">
<h3 data-number="13.2.3" class="anchored" data-anchor-id="bonus-daunting-snippet-but-that-makes-a-cool-figure-for-a-paper-modeling-pseudotime-dependent-gene-expression"><span class="header-section-number">13.2.3</span> BONUS: Daunting snippet but that makes a cool figure for a paper: modeling pseudotime-dependent gene expression</h3>
<p>Using pseudotime / cell, one can model gene expression along the differentiation process. This alleviates the need to study gene expression <strong>per cell</strong>, and allows one to focus on process-related effects (e.g.&nbsp;gene expression during a developmental trajectory).</p>
<ul>
<li><strong>Try to do so for few markers of spermatogonia, spermatocytes and spermatids.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'ID4'</span>, <span class="st">'SYCP3'</span>, <span class="st">'DMC1'</span>, <span class="st">'ACR'</span>, <span class="st">'PRM1'</span>, <span class="st">'PGK2'</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>fitExprs <span class="ot">&lt;-</span> <span class="fu">logcounts</span>(germcells[genes, ]) <span class="sc">|&gt;</span> <span class="co"># ----------------------------------- Get norm. counts for genes of interest</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.matrix</span>() <span class="sc">|&gt;</span> </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">t</span>() <span class="sc">|&gt;</span> </span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as_tibble</span>() <span class="sc">|&gt;</span> </span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(  <span class="co"># ----------------------------------------------------------------- Add information for each cell</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">cellID =</span> <span class="fu">colnames</span>(germcells), </span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">annotation =</span> <span class="fu">factor</span>(germcells<span class="sc">$</span>annotation, <span class="fu">c</span>(<span class="st">"Spermatogonia"</span>, <span class="st">"Spermatocytes"</span>, <span class="st">"Early spermatids"</span>, <span class="st">"Late spermatids"</span>)), </span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">pseudotime =</span> germcells<span class="sc">$</span>pseudotime</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">|&gt;</span> </span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="fu">contains</span>(genes), <span class="at">names_to =</span> <span class="st">'gene'</span>, <span class="at">values_to =</span> <span class="st">'obs_expr'</span>) <span class="sc">|&gt;</span> <span class="co"># - Pivot in "long" tidy format </span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">gene =</span> <span class="fu">factor</span>(gene, genes)) <span class="sc">|&gt;</span> </span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(gene) <span class="sc">|&gt;</span> <span class="co"># ------------------------------------------------------- Group rows by genes</span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">nest</span>(<span class="at">.key =</span> <span class="st">'data'</span>) <span class="sc">|&gt;</span> <span class="co"># -------------------------------------------------- For each gene, extract the subtable into a column named data</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">gamModel =</span> <span class="fu">map</span>(data, <span class="sc">~</span>mgcv<span class="sc">::</span><span class="fu">gam</span>(obs_expr <span class="sc">~</span> <span class="fu">s</span>(pseudotime, <span class="at">bs =</span> <span class="st">"cs"</span>), <span class="at">data =</span> .)), </span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">gamFitted_expr =</span> <span class="fu">map</span>(gamModel, predict) <span class="co"># ------------------------------ For each gene, fit the expression values ~ pseudotime with a GAM</span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">|&gt;</span> </span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span><span class="fu">ends_with</span>(<span class="st">'Model'</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unnest</span>(<span class="fu">c</span>(data, <span class="fu">ends_with</span>(<span class="st">'_expr'</span>))) <span class="co"># -------------------------------------- Unnest all the modelled expressions</span></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(fitExprs) <span class="sc">+</span> </span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> pseudotime, <span class="at">y =</span> obs_expr, <span class="at">col =</span> annotation), <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span> </span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x =</span> pseudotime, <span class="at">y =</span> gamFitted_expr), <span class="at">col =</span> <span class="st">'white'</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span> </span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x =</span> pseudotime, <span class="at">y =</span> gamFitted_expr), <span class="at">col =</span> <span class="st">'#af2d0c'</span>, <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_grid</span>(gene<span class="sc">~</span>., <span class="at">scales =</span> <span class="st">'free'</span>) <span class="sc">+</span> </span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">'logcounts'</span>) <span class="sc">+</span> </span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">'Fitted models of pseudotime-dependent gene expression'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
</section>
<section id="ordering-trajectory-with-rna-velocity" class="level2" data-number="13.3">
<h2 data-number="13.3" class="anchored" data-anchor-id="ordering-trajectory-with-rna-velocity"><span class="header-section-number">13.3</span> Ordering trajectory with RNA velocity</h2>
<p>As we saw earlier, TI does not necessarily know which direction is right for a given trajectory. This orientation can be sometimes refined using RNA velocity. Let’s see whether RNA velocity helps orientating our spermatocyte differentiation lineage trajectory here.</p>
<ul>
<li><strong>Read <code>velociraptor</code> documentation. What do you need to compute RNA velocity scores in R?</strong></li>
<li><strong>Import <code>spliced</code> and <code>unspliced</code> counts computed with <code>velocyto</code> in R.</strong></li>
<li><strong>Try and compute RNA velocity.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="do">## - Import entire GSE112013 dataset with spliced/unspliced counts </span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>full_GSE112013_counts <span class="ot">&lt;-</span> LoomExperiment<span class="sc">::</span><span class="fu">import</span>(<span class="st">'Share/Guo_testis/Guo_testis_full-counts.loom'</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="do">## - Filter `germcells` genes and cells to only retain those present in `full_GSE112013_counts`</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>germcells <span class="ot">&lt;-</span> germcells[</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rowData</span>(germcells)<span class="sc">$</span>gene_id <span class="sc">%in%</span> <span class="fu">rowData</span>(full_GSE112013_counts)<span class="sc">$</span>Accession, </span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    germcells<span class="sc">$</span>Barcode <span class="sc">%in%</span> full_GSE112013_counts<span class="sc">$</span>Barcode</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="do">## - Reorder rows of `full_GSE112013_counts_germcells` to match those of `germcells`</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>full_GSE112013_counts_germcells <span class="ot">&lt;-</span> full_GSE112013_counts[<span class="fu">match</span>(<span class="fu">rowData</span>(germcells)<span class="sc">$</span>gene_id, <span class="fu">rowData</span>(full_GSE112013_counts)<span class="sc">$</span>Accession), <span class="fu">match</span>(germcells<span class="sc">$</span>Barcode, full_GSE112013_counts<span class="sc">$</span>Barcode)]</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(germcells)</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(full_GSE112013_counts_germcells)</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="do">## - Add spliced/unspliced counts to germcells</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(germcells, <span class="st">'spliced'</span>, <span class="at">withDimnames=</span><span class="cn">FALSE</span>) <span class="ot">&lt;-</span> <span class="fu">assay</span>(full_GSE112013_counts_germcells, <span class="st">'spliced'</span>)</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(germcells, <span class="st">'unspliced'</span>, <span class="at">withDimnames=</span><span class="cn">FALSE</span>) <span class="ot">&lt;-</span> <span class="fu">assay</span>(full_GSE112013_counts_germcells, <span class="st">'unspliced'</span>)</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a><span class="do">## - Run velociraptor</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>velo_out <span class="ot">&lt;-</span> velociraptor<span class="sc">::</span><span class="fu">scvelo</span>(</span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>    germcells, </span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay.X =</span> <span class="st">"counts"</span>, </span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">subset.row =</span> <span class="fu">rowData</span>(germcells)<span class="sc">$</span>isHVG, </span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.dimred =</span> <span class="st">"corrected"</span></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>velo_out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>Embed the velocity field in tSNE scRNAseq embedding and plot the RNA velocity field on top of tSNE projection. Conclude.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>embedded_velo <span class="ot">&lt;-</span> velociraptor<span class="sc">::</span><span class="fu">embedVelocity</span>(<span class="fu">reducedDim</span>(germcells, <span class="st">"TSNE"</span>), velo_out)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(embedded_velo)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>grid_df <span class="ot">&lt;-</span> velociraptor<span class="sc">::</span><span class="fu">gridVectors</span>(<span class="fu">reducedDim</span>(germcells, <span class="st">"TSNE"</span>), embedded_velo, <span class="at">resolution =</span> <span class="dv">30</span>, <span class="at">scale =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(grid_df)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(germcells, <span class="st">'TSNE'</span>, <span class="at">colour_by =</span> <span class="st">"annotation"</span>, <span class="at">point_alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_segment</span>(</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">data =</span> grid_df, </span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> start<span class="fl">.1</span>, <span class="at">y =</span> start<span class="fl">.2</span>, <span class="at">xend =</span> end<span class="fl">.1</span>, <span class="at">yend =</span> end<span class="fl">.2</span>), </span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">arrow =</span> <span class="fu">arrow</span>(<span class="at">length =</span> <span class="fu">unit</span>(<span class="fl">0.05</span>, <span class="st">"inches"</span>), <span class="at">type =</span> <span class="st">"closed"</span>)</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="session-info" class="level2" data-number="13.4">
<h2 data-number="13.4" class="anchored" data-anchor-id="session-info"><span class="header-section-number">13.4</span> Session info</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.0 (2023-04-21)
 os       macOS Monterey 12.5.1
 system   aarch64, darwin20
 ui       X11
 language (EN)
 collate  en_GB.UTF-8
 ctype    en_GB.UTF-8
 tz       Europe/Paris
 date     2023-06-05
 pandoc   2.19.2 @ /opt/homebrew/bin/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 package                * version   date (UTC) lib source
 AnnotationDbi          * 1.61.0    2022-11-01 [1] Bioconductor
 AnnotationHub          * 3.7.3     2023-03-01 [1] Bioconductor
 Biobase                * 2.59.0    2022-11-01 [1] Bioconductor
 BiocFileCache          * 2.7.2     2023-02-17 [1] Bioconductor
 BiocGenerics           * 0.45.0    2022-11-01 [1] Bioconductor
 BiocManager            * 1.30.20   2023-02-24 [1] CRAN (R 4.3.0)
 BiocVersion              3.17.1    2022-12-20 [1] Bioconductor
 Biostrings               2.67.0    2022-11-01 [1] Bioconductor
 bit                      4.0.5     2022-11-15 [1] CRAN (R 4.3.0)
 bit64                    4.0.5     2020-08-30 [1] CRAN (R 4.3.0)
 bitops                   1.0-7     2021-04-24 [1] CRAN (R 4.3.0)
 blob                     1.2.3     2022-04-10 [1] CRAN (R 4.3.0)
 cachem                   1.0.7     2023-02-24 [1] CRAN (R 4.3.0)
 callr                    3.7.3     2022-11-02 [1] CRAN (R 4.3.0)
 cli                      3.6.0     2023-01-09 [1] CRAN (R 4.3.0)
 colorspace               2.1-0     2023-01-23 [1] CRAN (R 4.3.0)
 cowplot                * 1.1.1     2020-12-30 [1] CRAN (R 4.3.0)
 crayon                   1.5.2     2022-09-29 [1] CRAN (R 4.3.0)
 curl                     5.0.0     2023-01-12 [1] CRAN (R 4.3.0)
 DBI                      1.1.3     2022-06-18 [1] CRAN (R 4.3.0)
 dbplyr                 * 2.3.1     2023-02-24 [1] CRAN (R 4.3.0)
 devtools                 2.4.5     2022-10-11 [1] CRAN (R 4.3.0)
 digest                   0.6.31    2022-12-11 [1] CRAN (R 4.3.0)
 dplyr                  * 1.1.0     2023-01-29 [1] CRAN (R 4.3.0)
 ellipsis                 0.3.2     2021-04-29 [1] CRAN (R 4.3.0)
 evaluate                 0.20      2023-01-17 [1] CRAN (R 4.3.0)
 fansi                    1.0.4     2023-01-22 [1] CRAN (R 4.3.0)
 fastmap                  1.1.1     2023-02-24 [1] CRAN (R 4.3.0)
 filelock                 1.0.2     2018-10-05 [1] CRAN (R 4.3.0)
 forcats                * 1.0.0     2023-01-29 [1] CRAN (R 4.3.0)
 fs                       1.6.1     2023-02-06 [1] CRAN (R 4.3.0)
 generics                 0.1.3     2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb             1.35.15   2023-02-03 [1] Bioconductor
 GenomeInfoDbData         1.2.9     2022-11-04 [1] Bioconductor
 ggplot2                * 3.4.1     2023-02-10 [1] CRAN (R 4.3.0)
 glue                     1.6.2     2022-02-24 [1] CRAN (R 4.3.0)
 gtable                   0.3.1     2022-09-01 [1] CRAN (R 4.3.0)
 hms                      1.1.2     2022-08-19 [1] CRAN (R 4.3.0)
 htmltools                0.5.4     2022-12-07 [1] CRAN (R 4.3.0)
 htmlwidgets              1.6.1     2023-01-07 [1] CRAN (R 4.3.0)
 httpuv                   1.6.9     2023-02-14 [1] CRAN (R 4.3.0)
 httr                     1.4.5     2023-02-24 [1] CRAN (R 4.3.0)
 interactiveDisplayBase   1.37.0    2022-11-01 [1] Bioconductor
 IRanges                * 2.33.0    2022-11-01 [1] Bioconductor
 jsonlite                 1.8.4     2022-12-06 [1] CRAN (R 4.3.0)
 KEGGREST                 1.39.0    2022-11-01 [1] Bioconductor
 knitr                    1.42      2023-01-25 [1] CRAN (R 4.3.0)
 later                    1.3.0     2021-08-18 [1] CRAN (R 4.3.0)
 lattice                  0.21-8    2023-04-05 [2] CRAN (R 4.3.0)
 lifecycle                1.0.3     2022-10-07 [1] CRAN (R 4.3.0)
 lubridate              * 1.9.2     2023-02-10 [1] CRAN (R 4.3.0)
 magrittr                 2.0.3     2022-03-30 [1] CRAN (R 4.3.0)
 Matrix                   1.5-4     2023-04-04 [2] CRAN (R 4.3.0)
 memoise                  2.0.1     2021-11-26 [1] CRAN (R 4.3.0)
 mgcv                   * 1.8-42    2023-03-02 [2] CRAN (R 4.3.0)
 mime                     0.12      2021-09-28 [1] CRAN (R 4.3.0)
 miniUI                   0.1.1.1   2018-05-18 [1] CRAN (R 4.3.0)
 munsell                  0.5.0     2018-06-12 [1] CRAN (R 4.3.0)
 nlme                   * 3.1-162   2023-01-31 [2] CRAN (R 4.3.0)
 pillar                   1.8.1     2022-08-19 [1] CRAN (R 4.3.0)
 pkgbuild                 1.4.0     2022-11-27 [1] CRAN (R 4.3.0)
 pkgconfig                2.0.3     2019-09-22 [1] CRAN (R 4.3.0)
 pkgload                  1.3.2     2022-11-16 [1] CRAN (R 4.3.0)
 png                      0.1-8     2022-11-29 [1] CRAN (R 4.3.0)
 prettyunits              1.1.1     2020-01-24 [1] CRAN (R 4.3.0)
 processx                 3.8.0     2022-10-26 [1] CRAN (R 4.3.0)
 profvis                  0.3.7     2020-11-02 [1] CRAN (R 4.3.0)
 promises                 1.2.0.1   2021-02-11 [1] CRAN (R 4.3.0)
 ps                       1.7.2     2022-10-26 [1] CRAN (R 4.3.0)
 purrr                  * 1.0.1     2023-01-10 [1] CRAN (R 4.3.0)
 R6                       2.5.1     2021-08-19 [1] CRAN (R 4.3.0)
 rappdirs                 0.3.3     2021-01-31 [1] CRAN (R 4.3.0)
 Rcpp                     1.0.10    2023-01-22 [1] CRAN (R 4.3.0)
 RCurl                    1.98-1.10 2023-01-27 [1] CRAN (R 4.3.0)
 readr                  * 2.1.4     2023-02-10 [1] CRAN (R 4.3.0)
 remotes                  2.4.2     2021-11-30 [1] CRAN (R 4.3.0)
 rlang                    1.0.6     2022-09-24 [1] CRAN (R 4.3.0)
 rmarkdown                2.20      2023-01-19 [1] CRAN (R 4.3.0)
 RSQLite                  2.3.0     2023-02-17 [1] CRAN (R 4.3.0)
 S4Vectors              * 0.37.4    2023-02-26 [1] Bioconductor
 scales                   1.2.1     2022-08-20 [1] CRAN (R 4.3.0)
 sessioninfo              1.2.2     2021-12-06 [1] CRAN (R 4.3.0)
 shiny                    1.7.4     2022-12-15 [1] CRAN (R 4.3.0)
 stringi                  1.7.12    2023-01-11 [1] CRAN (R 4.3.0)
 stringr                * 1.5.0     2022-12-02 [1] CRAN (R 4.3.0)
 tibble                 * 3.1.8     2022-07-22 [1] CRAN (R 4.3.0)
 tidyr                  * 1.3.0     2023-01-24 [1] CRAN (R 4.3.0)
 tidyselect               1.2.0     2022-10-10 [1] CRAN (R 4.3.0)
 tidyverse              * 2.0.0     2023-02-22 [1] CRAN (R 4.3.0)
 timechange               0.2.0     2023-01-11 [1] CRAN (R 4.3.0)
 tzdb                     0.3.0     2022-03-28 [1] CRAN (R 4.3.0)
 urlchecker               1.0.1     2021-11-30 [1] CRAN (R 4.3.0)
 usethis                  2.1.6     2022-05-25 [1] CRAN (R 4.3.0)
 utf8                     1.2.3     2023-01-31 [1] CRAN (R 4.3.0)
 vctrs                    0.5.2     2023-01-23 [1] CRAN (R 4.3.0)
 withr                    2.5.0     2022-03-03 [1] CRAN (R 4.3.0)
 xfun                     0.37      2023-01-31 [1] CRAN (R 4.3.0)
 xtable                   1.8-4     2019-04-21 [1] CRAN (R 4.3.0)
 XVector                  0.39.0    2022-12-20 [1] Bioconductor
 yaml                     2.3.7     2023-01-23 [1] CRAN (R 4.3.0)
 zlibbioc                 1.45.0    2022-12-20 [1] Bioconductor

 [1] /Users/jacques/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../content/day4/Lecture6_pseudotime.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Lecture 6 - Trajectories and pseudotimes</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../content/day4/Lecture7_ATAC.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Lecture 7 - Advances in single-cell genomics: the epigenome</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">Single-cell RNAseq analysis with R/Bioconductor |<br>
J. Serizay O. Ashenberg</div>   
    <div class="nav-footer-right">This book was built with <a href="https://quarto.org/">Quarto</a>.</div>
  </div>
</footer>



</body></html>