<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.189">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="J. Serizay &amp; O. Ashenberg">
<title>Single-cell RNAseq analysis with R/Bioconductor - 15&nbsp; Lab 8 - Single-cell ATAC-seq analysis workflow</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../content/day5/Lecture8_spatial-transcriptomics.html" rel="next">
<link href="../../content/day4/Lecture7_ATAC.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }"><div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">
<span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Lab 8 - Single-cell ATAC-seq analysis workflow</span>
</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">Single-cell RNAseq analysis with R/Bioconductor</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/js2264/scRNAseq_Physalia_2023/" title="Source Code" class="sidebar-tool px-1"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">Welcome</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/program.html" class="sidebar-item-text sidebar-link">Program</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/Rstudio.html" class="sidebar-item-text sidebar-link">RStudio</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/prerequisites.html" class="sidebar-item-text sidebar-link">Prerequisites</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Day 1</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lecture1_introduction.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Lecture 1 - Introduction to scRNAseq analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lecture2_Bcl2matrix.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Lecture 2 - From sequencing reads to expression matrices</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lab1_Intro.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Lab 1: Familiarizing yourself with the course AWS instance</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lab2_processingreads.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Lab 2: From .bcl to count matrix</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">Day 2</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lecture3_qualitycontrol.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Lecture 3 - Quality control for scRNA-Seq data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lab3_Rbioc.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Lab 3: Introduction to R/Bioconductor</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lab4_data_wrangling_scRNAseq.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Lab 4 - Single-cell RNA-seq data wrangling</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">Day 3</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lecture4_clustering.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Lecture 4 - Identifying cell populations</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lab5_clustering.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Lab 5: Dimension reduction, clustering and annotation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lecture5_batchcorrection.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Lecture 5 - Data integration and batch effect correction</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lab6_batch_correction.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Lab 6 - Batch correction</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">Day 4</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lecture6_pseudotime.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Lecture 6 - Trajectories and pseudotimes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lab7_pseudotime.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Lab 7: Pseudotime analyses</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lecture7_ATAC.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Lecture 7 - Advances in single-cell genomics: the epigenome</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lab8_atac-seq.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Lab 8 - Single-cell ATAC-seq analysis workflow</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">Day 5</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day5/Lecture8_spatial-transcriptomics.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Lecture 8 - Advances in single-cell genomics: spatial transcriptomics</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/extra.html" class="sidebar-item-text sidebar-link">Extra resources</a>
  </div>
</li>
    </ul>
</div>
</nav><!-- margin-sidebar --><div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#process-human-pbmc-dataset" id="toc-process-human-pbmc-dataset" class="nav-link active" data-scroll-target="#process-human-pbmc-dataset"><span class="toc-section-number">15.1</span>  Process human PBMC dataset</a>
  <ul class="collapse">
<li><a href="#download-data" id="toc-download-data" class="nav-link" data-scroll-target="#download-data"><span class="toc-section-number">15.1.1</span>  Download data</a></li>
  <li><a href="#import-data" id="toc-import-data" class="nav-link" data-scroll-target="#import-data"><span class="toc-section-number">15.1.2</span>  Import data</a></li>
  <li><a href="#create-a-seurat-object" id="toc-create-a-seurat-object" class="nav-link" data-scroll-target="#create-a-seurat-object"><span class="toc-section-number">15.1.3</span>  Create a <code>Seurat</code> object</a></li>
  <li><a href="#check-qcs" id="toc-check-qcs" class="nav-link" data-scroll-target="#check-qcs"><span class="toc-section-number">15.1.4</span>  Check QCs</a></li>
  <li><a href="#filter-cells-and-features" id="toc-filter-cells-and-features" class="nav-link" data-scroll-target="#filter-cells-and-features"><span class="toc-section-number">15.1.5</span>  Filter cells and features</a></li>
  <li><a href="#dimensionality-reduction-and-clustering" id="toc-dimensionality-reduction-and-clustering" class="nav-link" data-scroll-target="#dimensionality-reduction-and-clustering"><span class="toc-section-number">15.1.6</span>  Dimensionality reduction and clustering</a></li>
  </ul>
</li>
  <li>
<a href="#chromvar-analysis" id="toc-chromvar-analysis" class="nav-link" data-scroll-target="#chromvar-analysis"><span class="toc-section-number">15.2</span>  chromVAR analysis</a>
  <ul class="collapse">
<li><a href="#get-a-summarizedexperiment-of-scatacseq-counts-over-peaks" id="toc-get-a-summarizedexperiment-of-scatacseq-counts-over-peaks" class="nav-link" data-scroll-target="#get-a-summarizedexperiment-of-scatacseq-counts-over-peaks"><span class="toc-section-number">15.2.1</span>  Get a SummarizedExperiment of scATACseq counts over peaks</a></li>
  <li><a href="#add-gc-bias-to-peaks" id="toc-add-gc-bias-to-peaks" class="nav-link" data-scroll-target="#add-gc-bias-to-peaks"><span class="toc-section-number">15.2.2</span>  Add GC bias to peaks</a></li>
  <li><a href="#map-motifs-over-peaks" id="toc-map-motifs-over-peaks" class="nav-link" data-scroll-target="#map-motifs-over-peaks"><span class="toc-section-number">15.2.3</span>  Map motifs over peaks</a></li>
  <li><a href="#search-for-motifs-with-high-deviation-of-mapping-compared-to-background" id="toc-search-for-motifs-with-high-deviation-of-mapping-compared-to-background" class="nav-link" data-scroll-target="#search-for-motifs-with-high-deviation-of-mapping-compared-to-background"><span class="toc-section-number">15.2.4</span>  Search for motifs with high deviation of mapping compared to background</a></li>
  <li><a href="#check-tf-motif-enrichment-in-different-cell-types" id="toc-check-tf-motif-enrichment-in-different-cell-types" class="nav-link" data-scroll-target="#check-tf-motif-enrichment-in-different-cell-types"><span class="toc-section-number">15.2.5</span>  Check TF motif enrichment in different cell types</a></li>
  </ul>
</li>
  <li><a href="#compute-gene-activity-scores" id="toc-compute-gene-activity-scores" class="nav-link" data-scroll-target="#compute-gene-activity-scores"><span class="toc-section-number">15.3</span>  Compute gene activity scores</a></li>
  <li><a href="#find-differentially-accessible-peaks" id="toc-find-differentially-accessible-peaks" class="nav-link" data-scroll-target="#find-differentially-accessible-peaks"><span class="toc-section-number">15.4</span>  Find differentially accessible peaks</a></li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/js2264/scRNAseq_Physalia_2023/edit/main/content/day4/Lab8_atac-seq.qmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/js2264/scRNAseq_Physalia_2023/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title d-none d-lg-block">
<span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Lab 8 - Single-cell ATAC-seq analysis workflow</span>
</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header><p>ATAC-seq data may be obtained in isolation using a single-cell ATAC-seq protocol (e.g.&nbsp;<a href="https://support.10xgenomics.com/single-cell-atac">10X scATACseq</a>) or in combination with gene expression data using a single-cell multiome protocole (e.g.&nbsp;<a href="https://www.10xgenomics.com/products/single-cell-multiome-atac-plus-gene-expression">10X multiome</a> and <a href="https://pubmed.ncbi.nlm.nih.gov/33098772/">SHARE-seq</a>).</p>
<p>Several packages are currently avaialble to process scATAC-seq data in R. These include <code>[Signac](https://satijalab.org/signac/index.html)</code> and <code>[ArchR](https://www.archrproject.com/bookdown/index.html#section)</code>. This lab will closely follow the <a href="https://satijalab.org/signac/articles/pbmc_vignette.html">processing steps</a> outlined in <code>Signac</code>, which interfaces well with <code>Seurat</code> for single-cell analysis.</p>
<p>In this lab, we will process a PBMC single-cell ATAC-seq (scATAC-seq) dataset and perform preliminary analysis to assess quality of these data. The data for this lab comes from <a href="https://www.10xgenomics.com/resources/datasets/10-k-peripheral-blood-mononuclear-cells-pbm-cs-from-a-healthy-donor-1-standard-1-0-1">10X Genomics</a>. The dataset contains roughly ~ 10,000 cells.</p>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Overarching goals
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>Import cells from human PBMC scATACseq dataset</li>
<li>Perform scATACseq quality controls and checks</li>
<li>Filter, normalize and plot PBMC scATACseq dataset</li>
<li>Run chromVAR to identify TF activity in each cell cluster<br>
</li>
<li>Compute gene activity scores and check known markers<br>
</li>
</ol>
</div>
</div>
<section id="process-human-pbmc-dataset" class="level2" data-number="15.1"><h2 data-number="15.1" class="anchored" data-anchor-id="process-human-pbmc-dataset">
<span class="header-section-number">15.1</span> Process human PBMC dataset</h2>
<section id="download-data" class="level3" data-number="15.1.1"><h3 data-number="15.1.1" class="anchored" data-anchor-id="download-data">
<span class="header-section-number">15.1.1</span> Download data</h3>
<ul>
<li><strong>Download the files related to scATACseq of all human PBMC cells.</strong></li>
</ul>
<p>Data comes from <a href="https://www.10xgenomics.com/resources/datasets/10-k-peripheral-blood-mononuclear-cells-pbm-cs-from-a-healthy-donor-1-standard-1-0-1">10X Genomics</a>.</p>
<p>Direct download links are provided below</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="st">'scATAC'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"https://cf.10xgenomics.com/samples/cell-atac/1.0.1/atac_v1_pbmc_10k/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5"</span>, <span class="st">"scATAC/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"https://cf.10xgenomics.com/samples/cell-atac/1.0.1/atac_v1_pbmc_10k/atac_v1_pbmc_10k_peaks.bed"</span>, <span class="st">"scATAC/atac_v1_pbmc_10k_peaks.bed"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"https://cf.10xgenomics.com/samples/cell-atac/1.0.1/atac_v1_pbmc_10k/atac_v1_pbmc_10k_singlecell.csv"</span>, <span class="st">"scATAC/atac_v1_pbmc_10k_singlecell.csv"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"https://cf.10xgenomics.com/samples/cell-atac/1.0.1/atac_v1_pbmc_10k/atac_v1_pbmc_10k_fragments.tsv.gz"</span>, <span class="st">"scATAC/atac_v1_pbmc_10k_fragments.tsv.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"https://cf.10xgenomics.com/samples/cell-atac/1.0.1/atac_v1_pbmc_10k/atac_v1_pbmc_10k_fragments.tsv.gz.tbi"</span>, <span class="st">"scATAC/atac_v1_pbmc_10k_fragments.tsv.gz.tbi"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="import-data" class="level3" data-number="15.1.2"><h3 data-number="15.1.2" class="anchored" data-anchor-id="import-data">
<span class="header-section-number">15.1.2</span> Import data</h3>
<p>Notice how the count matrix is in a <code>.h5</code> format. We have already encountered this format in <a href="../../content/day2/Lab3_Rbioc.html">Lab3</a>. Back then, we imported it with <code>DropletUtils::read10xCounts</code>.</p>
<ul>
<li><strong>Does this function work here?</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">DropletUtils</span><span class="fu">::</span><span class="fu">read10xCounts</span><span class="op">(</span><span class="st">"scATAC/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>This works because 10X Genomics make sure to distribute files in <code>.h5</code> format that are consistent accross single-cell sequencing mthodologies. However, the <code>SingleCellExperiment</code> obtained with this approach is not the most convenient, as it cannot natively leverage <code>fragments</code> file (see below).</p>
<p>Instead, we can create a <code>Signac</code> object, a flavour of <code>Seurat</code> objects.</p>
<ul>
<li><strong>Import counts matrix and feature annotations using an import function provided by <code>Seurat</code>.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac">Signac</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org">stringr</a></span><span class="op">)</span></span>
<span><span class="va">cnts</span> <span class="op">&lt;-</span> <span class="fu">Read10X_h5</span><span class="op">(</span><span class="st">'scATAC/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5'</span><span class="op">)</span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="st">'scATAC/atac_v1_pbmc_10k_peaks.bed'</span><span class="op">)</span></span>
<span><span class="va">features</span><span class="op">$</span><span class="va">peak</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html">str_replace</a></span><span class="op">(</span><span class="st">':'</span>, <span class="st">'-'</span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span></span>
<span>  file <span class="op">=</span> <span class="st">"scATAC/atac_v1_pbmc_10k_singlecell.csv"</span>,</span>
<span>  header <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  row.names <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>How many accessible genomic segments were found in this dataset?</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">features</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="create-a-seurat-object" class="level3" data-number="15.1.3"><h3 data-number="15.1.3" class="anchored" data-anchor-id="create-a-seurat-object">
<span class="header-section-number">15.1.3</span> Create a <code>Seurat</code> object</h3>
<p>The next step is to aggregate <code>counts</code> and <code>features</code> into a <code>ChromatinAssay</code>, a scATAC-seq flavour of <code>Seurat</code> standard <code>Assays</code>. The documentation for <code>?CreateChromatinAssay</code> indicates that the user can provide:</p>
<ol type="1">
<li>A <code>fragments</code> file, corresponding to the full list of all unique fragments mapped across all single cells.</li>
<li>Genomic annotations to the <code>ChromatinAssay</code>, corresponding to gene annotations, promoter positions, etc. Such annotations can be generated from Ensembl.</li>
</ol>
<ul>
<li><strong>Generate human annotations from Ensembl using a parsing function from <code>Seurat</code>.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Get human gene annotations (hg19/GRCh37) to feed it into the future `ChromatinAssay`</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">'EnsDb.Hsapiens.v75'</span><span class="op">)</span></span>
<span><span class="va">annotations</span> <span class="op">&lt;-</span> <span class="fu">GetGRangesFromEnsDb</span><span class="op">(</span>ensdb <span class="op">=</span> <span class="fu">EnsDb.Hsapiens.v75</span><span class="fu">::</span><span class="va">EnsDb.Hsapiens.v75</span><span class="op">)</span></span>
<span><span class="fu">seqlevelsStyle</span><span class="op">(</span><span class="va">annotations</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'UCSC'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>Create a <code>ChromatinAssay</code> using <code>counts</code>, <code>features</code>, <code>fragments</code> and <code>annotations</code>.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Create Chromatin Assay</span></span>
<span><span class="va">assay</span> <span class="op">&lt;-</span> <span class="fu">Signac</span><span class="fu">::</span><span class="fu">CreateChromatinAssay</span><span class="op">(</span></span>
<span>    counts <span class="op">=</span> <span class="va">cnts</span>,</span>
<span>    ranges <span class="op">=</span> <span class="va">features</span>,</span>
<span>    fragments <span class="op">=</span> <span class="st">"scATAC/atac_v1_pbmc_10k_fragments.tsv.gz"</span>,</span>
<span>    annotation <span class="op">=</span> <span class="va">annotations</span>,</span>
<span>    genome <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>    min.cells <span class="op">=</span> <span class="fl">10</span>, </span>
<span>    min.features <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span>
<span><span class="va">assay</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>What are the dimensions of this object? Are they comparable to the count matrix? Comment.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">assay</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>It’s finally time to wrap the <code>ChromatinAssay</code> into a <code>Seurat</code> standard object. This is done using the <code>CreateSeuratObject</code>, as already covered in <a href="../../content/day3/Lab6_batch_correction.html">Lab6</a></p>
<ul>
<li><strong>Create a <code>PBMC</code> <code>Seurat</code> object.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Create Seurat object</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu">CreateSeuratObject</span><span class="op">(</span></span>
<span>    counts <span class="op">=</span> <span class="va">assay</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">'ATAC'</span>, </span>
<span>    meta.data <span class="op">=</span> <span class="va">metadata</span><span class="op">[</span><span class="va">metadata</span><span class="op">$</span><span class="va">is__cell_barcode</span> <span class="op">==</span> <span class="fl">1</span>, <span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="va">PBMC</span></span>
<span></span>
<span><span class="va">PBMC</span><span class="op">[[</span><span class="st">'ATAC'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html">granges</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span></span>
<span><span class="fu">Annotation</span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="check-qcs" class="level3" data-number="15.1.4"><h3 data-number="15.1.4" class="anchored" data-anchor-id="check-qcs">
<span class="header-section-number">15.1.4</span> Check QCs</h3>
<section id="cell-based-qcs" class="level4" data-number="15.1.4.1"><h4 data-number="15.1.4.1" class="anchored" data-anchor-id="cell-based-qcs">
<span class="header-section-number">15.1.4.1</span> Cell-based QCs</h4>
<p>The fraction of reads in peaks (FRiP) is a good indicator of how well each cell was handled during scATACseq processing.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">PBMC</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">FRiP</span> <span class="op">&lt;-</span> <span class="va">PBMC</span><span class="op">$</span><span class="va">peak_region_fragments</span> <span class="op">/</span> <span class="va">PBMC</span><span class="op">$</span><span class="va">passed_filters</span></span>
<span><span class="va">PBMC</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">nCount_ATAC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="fu">GetAssayData</span><span class="op">(</span><span class="va">PBMC</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">PBMC</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">nFeature_ATAC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="fu">GetAssayData</span><span class="op">(</span><span class="va">PBMC</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">$</span><span class="va">FRiP</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">$</span><span class="va">nCount_ATAC</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">$</span><span class="va">nFeature_ATAC</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="peaks-based-qcs" class="level4" data-number="15.1.4.2"><h4 data-number="15.1.4.2" class="anchored" data-anchor-id="peaks-based-qcs">
<span class="header-section-number">15.1.4.2</span> Peaks-based QCs</h4>
<ul>
<li><strong>Which analysis are the <code>fragments</code> required for, exactly?</strong></li>
<li><strong>Could we still perform normalization/clustering/annotation without them? And motif enrichment analysis?</strong></li>
</ul>
<p>Since we do have the <code>fragments</code> file at hand, most of the QC steps are available (e.g.&nbsp;<code>TSSEnrichment</code>, <code>NucleosomeSignal</code> or fragment size distribution). Let’s go through them one by one.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># compute nucleosome signal score per cell</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu">NucleosomeSignal</span><span class="op">(</span>object <span class="op">=</span> <span class="va">PBMC</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compute TSS enrichment score per cell</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu">Signac</span><span class="fu">::</span><span class="fu">TSSEnrichment</span><span class="op">(</span>object <span class="op">=</span> <span class="va">PBMC</span>, fast <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>The <code>TSSPlot</code> function from <code>Signac</code> can be used to plot the fragment count per peak ~ TSS enrichment.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">PBMC</span><span class="op">$</span><span class="va">high.tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">$</span><span class="va">TSS.enrichment</span> <span class="op">&gt;</span> <span class="fl">3.5</span>, <span class="st">'High'</span>, <span class="st">'Low'</span><span class="op">)</span></span>
<span><span class="fu">TSSPlot</span><span class="op">(</span><span class="va">PBMC</span>, group.by <span class="op">=</span> <span class="st">'high.tss'</span><span class="op">)</span> <span class="op">+</span> <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">PBMC</span><span class="op">$</span><span class="va">high.tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">$</span><span class="va">TSS.enrichment</span> <span class="op">&gt;</span> <span class="fl">2.5</span>, <span class="st">'High'</span>, <span class="st">'Low'</span><span class="op">)</span></span>
<span><span class="fu">TSSPlot</span><span class="op">(</span><span class="va">PBMC</span>, group.by <span class="op">=</span> <span class="st">'high.tss'</span><span class="op">)</span> <span class="op">+</span> <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>The <code>FragmentHistogram</code> function from <code>Signac</code> can be used to plot the fragment size distribution in peaks with different nucleosome signals.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">PBMC</span><span class="op">$</span><span class="va">nucleosome_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">$</span><span class="va">nucleosome_signal</span> <span class="op">&gt;</span> <span class="fl">4</span>, <span class="st">'NS &gt; 4'</span>, <span class="st">'NS &lt; 4'</span><span class="op">)</span></span>
<span><span class="fu">FragmentHistogram</span><span class="op">(</span>object <span class="op">=</span> <span class="va">PBMC</span>, group.by <span class="op">=</span> <span class="st">'nucleosome_group'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>The new <code>DensityScatter</code> function from <code>Signac</code> can be used to plot the fragment count per peak ~ TSS enrichment, combining both metrics into a single plot.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">DensityScatter</span><span class="op">(</span><span class="va">PBMC</span>, x <span class="op">=</span> <span class="st">'TSS.enrichment'</span>, y <span class="op">=</span> <span class="st">'nucleosome_signal'</span>, log_x <span class="op">=</span> <span class="cn">FALSE</span>, quantiles <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section></section><section id="filter-cells-and-features" class="level3" data-number="15.1.5"><h3 data-number="15.1.5" class="anchored" data-anchor-id="filter-cells-and-features">
<span class="header-section-number">15.1.5</span> Filter cells and features</h3>
<p>We do have other pre-computed metrics (e.g.&nbsp;<code>FRIP</code> and <code>depth</code>).</p>
<ul>
<li><strong>Check these metrics and filter the <code>Seurat</code> object (cells and features) as deemed appropriate.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Filter data </span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">PBMC</span>, subset <span class="op">=</span> <span class="va">nCount_ATAC</span> <span class="op">&gt;</span> <span class="fl">3000</span> <span class="op">&amp;</span> <span class="va">nCount_ATAC</span> <span class="op">&lt;</span> <span class="fl">100000</span><span class="op">)</span> </span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">PBMC</span>, subset <span class="op">=</span> <span class="va">nFeature_ATAC</span> <span class="op">&gt;</span> <span class="fl">1000</span> <span class="op">&amp;</span> <span class="va">nFeature_ATAC</span> <span class="op">&lt;</span> <span class="fl">20000</span><span class="op">)</span> </span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">PBMC</span>, subset <span class="op">=</span> <span class="va">FRIP</span> <span class="op">&gt;</span> <span class="fl">0.30</span><span class="op">)</span> </span>
<span><span class="co">## - Remove peaks with low coverage</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="va">PBMC</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu">GetAssayData</span><span class="op">(</span><span class="va">PBMC</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">10</span>, <span class="op">]</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="va">PBMC</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu">GetAssayData</span><span class="op">(</span><span class="va">PBMC</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">10</span>, <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="dimensionality-reduction-and-clustering" class="level3" data-number="15.1.6"><h3 data-number="15.1.6" class="anchored" data-anchor-id="dimensionality-reduction-and-clustering">
<span class="header-section-number">15.1.6</span> Dimensionality reduction and clustering</h3>
<ul>
<li><strong>Now that the dataset is filtered, normalize (by using TF-IDF approach) then further reduce the dimensionality for visualization purposes.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Normalize data </span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu">RunTFIDF</span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span> </span>
<span></span>
<span><span class="co">## - Reduce dimensionality</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu">FindTopFeatures</span><span class="op">(</span><span class="va">PBMC</span>, min.cutoff <span class="op">=</span> <span class="st">'q50'</span><span class="op">)</span> </span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu">RunSVD</span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span> </span>
<span></span>
<span><span class="co">## - Label clusters according to the original publication</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu">RunUMAP</span><span class="op">(</span>object <span class="op">=</span> <span class="va">PBMC</span>, reduction <span class="op">=</span> <span class="st">'lsi'</span>, dims <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu">FindNeighbors</span><span class="op">(</span>object <span class="op">=</span> <span class="va">PBMC</span>, reduction <span class="op">=</span> <span class="st">'lsi'</span>, dims <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu">FindClusters</span><span class="op">(</span>object <span class="op">=</span> <span class="va">PBMC</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, algorithm <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## - Visualize data </span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">PBMC</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>What can you observe in the <code>UMAP</code> projection of the dataset? Comment on the separation of some cell types into different spatially-resolved clusters.</strong></li>
</ul></section></section><section id="chromvar-analysis" class="level2" data-number="15.2"><h2 data-number="15.2" class="anchored" data-anchor-id="chromvar-analysis">
<span class="header-section-number">15.2</span> chromVAR analysis</h2>
<section id="get-a-summarizedexperiment-of-scatacseq-counts-over-peaks" class="level3" data-number="15.2.1"><h3 data-number="15.2.1" class="anchored" data-anchor-id="get-a-summarizedexperiment-of-scatacseq-counts-over-peaks">
<span class="header-section-number">15.2.1</span> Get a SummarizedExperiment of scATACseq counts over peaks</h3>
<p><code>chromVAR</code> works with raw counts stored as a <code>RangedSummarizedExperiment</code>. By now, you should be able to easily extract the raw counts from a specific <code>Seurat</code> assay. Store it in a <code>RangedSummarizedExperiment</code> (~ equivalent to <code>Seurat</code> assay but in <code>Bioconductor</code>).</p>
<ul>
<li><strong>What are the dimensions of the <code>RangedSummarizedExperiment</code> generated? What are the features and what are the columns?</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="va">sumExp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">SummarizedExperiment</a></span><span class="op">(</span></span>
<span>    assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">'counts'</span> <span class="op">=</span> <span class="fu">GetAssayData</span><span class="op">(</span><span class="va">PBMC</span>, assay <span class="op">=</span> <span class="st">'ATAC'</span>, slot <span class="op">=</span> <span class="st">'counts'</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    rowRanges <span class="op">=</span> <span class="fu">GetAssayData</span><span class="op">(</span><span class="va">PBMC</span>, assay <span class="op">=</span> <span class="st">'ATAC'</span>, slot <span class="op">=</span> <span class="st">'ranges'</span><span class="op">)</span>, </span>
<span>    colData <span class="op">=</span> <span class="va">PBMC</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">sumExp</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="add-gc-bias-to-peaks" class="level3" data-number="15.2.2"><h3 data-number="15.2.2" class="anchored" data-anchor-id="add-gc-bias-to-peaks">
<span class="header-section-number">15.2.2</span> Add GC bias to peaks</h3>
<p>This step is important to correct GC bias associated with each peak.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sumExp</span> <span class="op">&lt;-</span> <span class="fu">chromVAR</span><span class="fu">::</span><span class="fu">addGCBias</span><span class="op">(</span><span class="va">sumExp</span>, genome <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va">BSgenome.Hsapiens.UCSC.hg19</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="map-motifs-over-peaks" class="level3" data-number="15.2.3"><h3 data-number="15.2.3" class="anchored" data-anchor-id="map-motifs-over-peaks">
<span class="header-section-number">15.2.3</span> Map motifs over peaks</h3>
<p>Using <code>motifmatchr</code> package, one can map a list of motifs (e.g.&nbsp;from JASPAR database) over a <code>RangedSummarizedExperiment</code> (or anything which can be coerced to a <code>GRanges</code>, actually).</p>
<ul>
<li><strong>First, you need to import motifs from a public database in <code>R</code>. This can be done (among other ways) with the <code>TFBSTools</code> pacakge.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Import motifs from JASPAR database</span></span>
<span><span class="va">motifs</span> <span class="op">&lt;-</span> <span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/getMatrixSet-methods.html">getMatrixSet</a></span><span class="op">(</span></span>
<span>    <span class="fu">JASPAR2020</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/JASPAR2020/man/JASPAR2020-class.html">JASPAR2020</a></span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>species <span class="op">=</span> <span class="fl">9606</span>, all_versions <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">motifs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html">name</a></span><span class="op">(</span><span class="va">motifs</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><p><strong>Now map the subset of motifs of interest to the filtered peaks, using <code>motifmatchr::matchMotifs()</code>.</strong></p></li>
<li><p><strong>Read <code>matchMotifs()</code> documentation and run it on your <code>sumExp</code>.</strong></p></li>
<li><p><strong>What is the class of the generated object? Its dimensions?</strong></p></li>
<li><p><strong>What are the features and what are the columns?</strong></p></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">16</span>, progressbar <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">motif_mappings</span> <span class="op">&lt;-</span> <span class="fu">motifmatchr</span><span class="fu">::</span><span class="fu">matchMotifs</span><span class="op">(</span><span class="va">motifs</span>, <span class="va">sumExp</span>, genome <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va">BSgenome.Hsapiens.UCSC.hg19</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">motif_mappingss</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">motif_mappings</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>Because there are so many motifs and so many peaks, and in interest of time, we will perform <code>chromVAR</code> analysis on a subset of motifs only. You can manually filter the motifs list to motifs that may be relevant in our context (e.g.&nbsp;<code>EOMES</code>, or <code>GATA2</code>, …).</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Filter to interesting motifs to keep analysis relatively quick</span></span>
<span><span class="co">#       Feel free to select any TF of interest!</span></span>
<span><span class="va">motif_mappings_sub</span> <span class="op">&lt;-</span> <span class="va">motif_mappings</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'EOMES'</span>, <span class="st">'GATA2'</span>, <span class="st">'IRF8'</span>, <span class="st">'EBF1'</span>, <span class="st">'SPI1'</span>, <span class="st">'TCF3'</span>, <span class="st">'CEBPA'</span><span class="op">)</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="search-for-motifs-with-high-deviation-of-mapping-compared-to-background" class="level3" data-number="15.2.4"><h3 data-number="15.2.4" class="anchored" data-anchor-id="search-for-motifs-with-high-deviation-of-mapping-compared-to-background">
<span class="header-section-number">15.2.4</span> Search for motifs with high deviation of mapping compared to background</h3>
<p><code>chromVAR</code>’s <code>computeDeviations()</code> function combines (1) the peak counts / cell (stored here in <code>sumExp</code>) and (2) the TF mapping over each peak (stored here in <code>motif_mappings</code>). to assess the mapping deviation for each TF over each cell compared to other cells.</p>
<ul>
<li><strong>Run <code>computeDeviations()</code> on the set of peaks and motifs.</strong></li>
<li><strong>What is the class of the generated object? Its dimensions?</strong></li>
<li><strong>What are the features and what are the columns?</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Find background signal</span></span>
<span><span class="va">bg</span> <span class="op">&lt;-</span> <span class="fu">chromVAR</span><span class="fu">::</span><span class="fu">getBackgroundPeaks</span><span class="op">(</span>object <span class="op">=</span> <span class="va">sumExp</span><span class="op">)</span></span>
<span><span class="co">## - For each motif, compute its mapping deviation over the filtered peaks</span></span>
<span><span class="va">motif_deviations</span> <span class="op">&lt;-</span> <span class="fu">chromVAR</span><span class="fu">::</span><span class="fu">computeDeviations</span><span class="op">(</span>object <span class="op">=</span> <span class="va">sumExp</span>, annotations <span class="op">=</span> <span class="va">motif_mappings_sub</span>, background_peaks <span class="op">=</span> <span class="va">bg</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="check-tf-motif-enrichment-in-different-cell-types" class="level3" data-number="15.2.5"><h3 data-number="15.2.5" class="anchored" data-anchor-id="check-tf-motif-enrichment-in-different-cell-types">
<span class="header-section-number">15.2.5</span> Check TF motif enrichment in different cell types</h3>
<p>The <code>motif_deviations</code> object can be added as a new assay to the <code>PBMC</code> <code>Seurat</code> object. This way, one can rely on <code>Seurat</code>-based plotting functions to plot cells in their preferred dimensional space, and color them using motif deviation scores computed with <code>chromVAR</code>.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">PBMC</span><span class="op">[[</span><span class="st">'MOTIF'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu">CreateAssayObject</span><span class="op">(</span>counts <span class="op">=</span> <span class="fu">chromVAR</span><span class="fu">::</span><span class="fu">deviationScores</span><span class="op">(</span><span class="va">motif_deviations</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'MOTIF'</span></span>
<span><span class="va">list_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">motif</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">FeaturePlot</span><span class="op">(</span><span class="va">PBMC</span>, features <span class="op">=</span> <span class="va">motif</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>        <span class="fu">scale_colour_gradientn</span><span class="op">(</span></span>
<span>            colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'#190886'</span>, <span class="st">'#6F07F8'</span>, <span class="st">'#F954A5'</span>, <span class="st">'#FF9D66'</span>, <span class="st">'#edf118'</span><span class="op">)</span>,</span>
<span>            limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span>, oob <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/oob.html">squish</a></span></span>
<span>        <span class="op">)</span> <span class="op">+</span> </span>
<span>        <span class="fu">theme</span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">list_p</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">list_p</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">PBMC</span>, group.by <span class="op">=</span> <span class="st">'renamed_clusters'</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">list_p</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section></section><section id="compute-gene-activity-scores" class="level2" data-number="15.3"><h2 data-number="15.3" class="anchored" data-anchor-id="compute-gene-activity-scores">
<span class="header-section-number">15.3</span> Compute gene activity scores</h2>
<p>Signac’s <code>GeneActivity()</code> function require scATACseq fragment information. Since we have them, we can estimate a gene activity score for each gene in the <code>annotations</code>.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene.activities</span> <span class="op">&lt;-</span> <span class="fu">GeneActivity</span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>We can now save this new object as an <code>Assay</code> in the <code>PBMC</code> object and normalize it.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">PBMC</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">CreateAssayObject</span><span class="op">(</span>counts <span class="op">=</span> <span class="va">gene.activities</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># - Normalize the new RNA assay, this time with `SCTransform`</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu">SCTransform</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">PBMC</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">'RNA'</span>,</span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">PBMC</span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SCT"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>One can now perform “gene differential expression”-like analysis using the <code>SCT</code> assay!</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Reduce dimensionality to visualize cells in 2D</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">RunUMAP</span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">'pca'</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## - Plot gene expression </span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">FeaturePlot</span><span class="op">(</span><span class="va">PBMC</span>, features <span class="op">=</span> <span class="st">'MS4A1'</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span> <span class="co"># Or NCR1 (NKp46), CD8A, CD14, MS4A1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>Now, we can leverage this to annotate clusters according to known PBMC markers</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Reduce dimensionality to visualize cells in 2D</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">RunUMAP</span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">'pca'</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## - Plot gene expression </span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">FeaturePlot</span><span class="op">(</span><span class="va">PBMC</span>, features <span class="op">=</span> <span class="st">'MS4A1'</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span> <span class="co"># Or NCR1 (NKp46), CD8A, CD14, MS4A1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="find-differentially-accessible-peaks" class="level2" data-number="15.4"><h2 data-number="15.4" class="anchored" data-anchor-id="find-differentially-accessible-peaks">
<span class="header-section-number">15.4</span> Find differentially accessible peaks</h2>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># change back to working with peaks instead of gene activities</span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'peaks'</span></span>
<span></span>
<span><span class="va">da_peaks</span> <span class="op">&lt;-</span> <span class="fu">FindMarkers</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  ident.1 <span class="op">=</span> <span class="st">"CD4 Naive"</span>,</span>
<span>  ident.2 <span class="op">=</span> <span class="st">"CD14+ Monocytes"</span>,</span>
<span>  test.use <span class="op">=</span> <span class="st">'LR'</span>,</span>
<span>  latent.vars <span class="op">=</span> <span class="st">'nCount_peaks'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu">VlnPlot</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">da_peaks</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  pt.size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  idents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD4 Naive"</span>,<span class="st">"CD14+ Monocytes"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu">FeaturePlot</span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">da_peaks</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  pt.size <span class="op">=</span> <span class="fl">0.1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot1</span> <span class="op">|</span> <span class="va">plot2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../content/day4/Lecture7_ATAC.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Lecture 7 - Advances in single-cell genomics: the epigenome</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../content/day5/Lecture8_spatial-transcriptomics.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Lecture 8 - Advances in single-cell genomics: spatial transcriptomics</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">Single-cell RNAseq analysis with R/Bioconductor |<br>
J. Serizay O. Ashenberg</div>   
    <div class="nav-footer-right">This book was built with <a href="https://quarto.org/">Quarto</a>.</div>
  </div>
</footer>


</body></html>