<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.189">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="J. Serizay &amp; O. Ashenberg">
<title>Single-cell RNAseq analysis with R/Bioconductor - 15&nbsp; Lab 8 - Single-cell ATAC-seq analysis workflow</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../content/day5/Lecture8_spatial-transcriptomics.html" rel="next">
<link href="../../content/day4/Lecture7_ATAC.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }"><div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">
<span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Lab 8 - Single-cell ATAC-seq analysis workflow</span>
</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">Single-cell RNAseq analysis with R/Bioconductor</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/js2264/scRNAseq_Physalia_2023/" title="Source Code" class="sidebar-tool px-1"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">Welcome</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/program.html" class="sidebar-item-text sidebar-link">Program</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/Rstudio.html" class="sidebar-item-text sidebar-link">RStudio</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/prerequisites.html" class="sidebar-item-text sidebar-link">Prerequisites</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Day 1</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lecture1_introduction.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Lecture 1 - Introduction to scRNAseq analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lab1_Intro.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Lab 1: Familiarizing yourself with the course AWS instance</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lecture2_Bcl2matrix.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Lecture 2 - From sequencing reads to expression matrices</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lab2_processingreads.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Lab 2: From .bcl to count matrix</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">Day 2</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lab3_Rbioc.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Lab 3: Introduction to R/Bioconductor</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lecture3_qualitycontrol.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Lecture 3 - Quality control for scRNA-Seq data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lab4_data_wrangling_scRNAseq.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Lab 4 - Single-cell RNA-seq data wrangling</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">Day 3</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lecture4_clustering.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Lecture 4 - Identifying cell populations</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lab5_clustering.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Lab 5: Dimension reduction, clustering and annotation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lecture5_batchcorrection.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Lecture 5 - Data integration and batch effect correction</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lab6_batch_correction.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Lab 6 - Batch correction</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">Day 4</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lecture6_pseudotime.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Lecture 6 - Trajectories and pseudotimes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lab7_pseudotime.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Lab 7: Pseudotime analyses</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lecture7_ATAC.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Lecture 7 - Advances in single-cell genomics: the epigenome</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lab8_atac-seq.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Lab 8 - Single-cell ATAC-seq analysis workflow</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">Day 5</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day5/Lecture8_spatial-transcriptomics.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Lecture 8 - Advances in single-cell genomics: spatial transcriptomics</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/extra.html" class="sidebar-item-text sidebar-link">Extra resources</a>
  </div>
</li>
    </ul>
</div>
</nav><!-- margin-sidebar --><div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#process-human-hematopoiesis-dataset" id="toc-process-human-hematopoiesis-dataset" class="nav-link active" data-scroll-target="#process-human-hematopoiesis-dataset"><span class="toc-section-number">15.1</span>  Process human hematopoiesis dataset</a>
  <ul class="collapse">
<li><a href="#download-data" id="toc-download-data" class="nav-link" data-scroll-target="#download-data"><span class="toc-section-number">15.1.1</span>  Download data</a></li>
  <li><a href="#process-data-with-signac" id="toc-process-data-with-signac" class="nav-link" data-scroll-target="#process-data-with-signac"><span class="toc-section-number">15.1.2</span>  Process data with Signac</a></li>
  </ul>
</li>
  <li>
<a href="#chromvar-analysis" id="toc-chromvar-analysis" class="nav-link" data-scroll-target="#chromvar-analysis"><span class="toc-section-number">15.2</span>  chromVAR analysis</a>
  <ul class="collapse">
<li><a href="#get-a-summarizedexperiment-of-scatacseq-counts-over-peaks" id="toc-get-a-summarizedexperiment-of-scatacseq-counts-over-peaks" class="nav-link" data-scroll-target="#get-a-summarizedexperiment-of-scatacseq-counts-over-peaks"><span class="toc-section-number">15.2.1</span>  Get a SummarizedExperiment of scATACseq counts over peaks</a></li>
  <li><a href="#add-gc-bias-to-peaks" id="toc-add-gc-bias-to-peaks" class="nav-link" data-scroll-target="#add-gc-bias-to-peaks"><span class="toc-section-number">15.2.2</span>  Add GC bias to peaks</a></li>
  <li><a href="#map-motifs-over-peaks" id="toc-map-motifs-over-peaks" class="nav-link" data-scroll-target="#map-motifs-over-peaks"><span class="toc-section-number">15.2.3</span>  Map motifs over peaks</a></li>
  <li><a href="#search-for-motifs-with-high-deviation-of-mapping-compared-to-background" id="toc-search-for-motifs-with-high-deviation-of-mapping-compared-to-background" class="nav-link" data-scroll-target="#search-for-motifs-with-high-deviation-of-mapping-compared-to-background"><span class="toc-section-number">15.2.4</span>  Search for motifs with high deviation of mapping compared to background</a></li>
  <li><a href="#check-tf-motif-enrichment-in-different-cell-types" id="toc-check-tf-motif-enrichment-in-different-cell-types" class="nav-link" data-scroll-target="#check-tf-motif-enrichment-in-different-cell-types"><span class="toc-section-number">15.2.5</span>  Check TF motif enrichment in different cell types</a></li>
  </ul>
</li>
  <li>
<a href="#compute-gene-activity-scores" id="toc-compute-gene-activity-scores" class="nav-link" data-scroll-target="#compute-gene-activity-scores"><span class="toc-section-number">15.3</span>  Compute gene activity scores</a>
  <ul class="collapse">
<li><a href="#manually-calculate-gene-activity-scores" id="toc-manually-calculate-gene-activity-scores" class="nav-link" data-scroll-target="#manually-calculate-gene-activity-scores"><span class="toc-section-number">15.3.1</span>  Manually calculate gene activity scores</a></li>
  <li><a href="#compare-gene-activity-and-tf-activity" id="toc-compare-gene-activity-and-tf-activity" class="nav-link" data-scroll-target="#compare-gene-activity-and-tf-activity"><span class="toc-section-number">15.3.2</span>  Compare gene activity and TF activity</a></li>
  </ul>
</li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/js2264/scRNAseq_Physalia_2023/edit/main/content/day4/Lab8_atac-seq.qmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/js2264/scRNAseq_Physalia_2023/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title d-none d-lg-block">
<span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Lab 8 - Single-cell ATAC-seq analysis workflow</span>
</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header><p>In this lab, we will process a single-cell ATAC-seq (scATAC-seq) dataset and perform preliminary analysis to assess quality of these data. The data for this lab comes from <a href="https://pubmed.ncbi.nlm.nih.gov/31375813/">Satpathy et al., Nature Biotechnology, 2019</a>. Counts data used in this lab can be found in GEO (GSE: GSE129785).</p>
<p>ATAC-seq data may be obtained in isolation using a single-cell ATAC-seq protocol (e.g.&nbsp;<a href="https://support.10xgenomics.com/single-cell-atac">10X scATACseq</a>) or in combination with gene expression data using a single-cell multiome protocole (e.g.&nbsp;<a href="https://www.10xgenomics.com/products/single-cell-multiome-atac-plus-gene-expression">10X multiome</a> and <a href="https://pubmed.ncbi.nlm.nih.gov/33098772/">SHARE-seq</a>).</p>
<p>Several packages are currently avaialble to process scATAC-seq data in R. These include <a href="https://satijalab.org/signac/index.html">Signac</a> and <a href="https://www.archrproject.com/bookdown/index.html#section">ArchR</a>. This lab will closely follow the <a href="https://satijalab.org/signac/articles/pbmc_vignette.html">processing steps</a> outlined in Signac, which interfaces well with Seurat for single-cell analysis.</p>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Overarching goals
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>Import, filter, normalize and plot cells from human hematopoiesis scATACseq dataset</li>
<li>Run chromVAR to identify TF activity in each cell cluster<br>
</li>
<li>Compute gene activity scores and check known markers<br>
</li>
</ol>
</div>
</div>
<section id="process-human-hematopoiesis-dataset" class="level2" data-number="15.1"><h2 data-number="15.1" class="anchored" data-anchor-id="process-human-hematopoiesis-dataset">
<span class="header-section-number">15.1</span> Process human hematopoiesis dataset</h2>
<section id="download-data" class="level3" data-number="15.1.1"><h3 data-number="15.1.1" class="anchored" data-anchor-id="download-data">
<span class="header-section-number">15.1.1</span> Download data</h3>
<p>The data comes from Satpathy et al., Nat. Biotechnol. 2019 (<a href="https://doi.org/10.1038/s41587-019-0206-z">DOI: 10.1038/s41587-019-0206-z</a>). Counts data is available from GEO (GSE: <code>GSE129785</code>).</p>
<ul>
<li><strong>Download the files related to scATACseq of all human hematopoiesis cells.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="st">'scATAC'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">'https://ftp.ncbi.nlm.nih.gov/geo/series/GSE129nnn/GSE129785/suppl/GSE129785_scATAC-Hematopoiesis-All.cell_barcodes.txt.gz'</span>, <span class="st">'scATAC/barcodes.tsv.gz'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">'https://ftp.ncbi.nlm.nih.gov/geo/series/GSE129nnn/GSE129785/suppl/GSE129785_scATAC-Hematopoiesis-All.mtx.gz'</span>, <span class="st">'scATAC/matrix.mtx.gz'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">'https://ftp.ncbi.nlm.nih.gov/geo/series/GSE129nnn/GSE129785/suppl/GSE129785_scATAC-Hematopoiesis-All.peaks.txt.gz'</span>, <span class="st">'scATAC/features.tsv.gz'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="process-data-with-signac" class="level3" data-number="15.1.2"><h3 data-number="15.1.2" class="anchored" data-anchor-id="process-data-with-signac">
<span class="header-section-number">15.1.2</span> Process data with Signac</h3>
<section id="import-data" class="level4" data-number="15.1.2.1"><h4 data-number="15.1.2.1" class="anchored" data-anchor-id="import-data">
<span class="header-section-number">15.1.2.1</span> Import data</h4>
<p>Notice how the count matrix is in a different format than previously seen. You will frequently encounter this type of situation. Analysis packages often provide import functions for the different input formats. Check <code>Seurat</code> documentation if needed, to identify which function can be used to import a <code>mtx</code> format.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cnts</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu">Read10X</span><span class="op">(</span><span class="st">'scATAC/'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>Does it work? Why? The most likely reason is that the count matrix is not structured exactly like the one which comes straight out of a <code>cellranger count</code> pipeline. It is a very frequent issue occurring when one attempts to analyse public data.</p>
<p>The alternative, longer route is to import each file separately, make sure they are compatible with each other, and merge them manually into a <code>SeuratChromatinAssay</code>.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">## - Load required libraries </span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plyranges)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Signac)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="do">## - Import counts, barcodes and features separately</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>barcode <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">'scATAC/'</span>, <span class="st">"barcodes.tsv.gz"</span>)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>features <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">'scATAC/'</span>, <span class="st">"features.tsv.gz"</span>)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>matrix <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">'scATAC/'</span>, <span class="st">"matrix.mtx.gz"</span>)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>cnts <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">readMM</span>(<span class="st">'scATAC/matrix.mtx.gz'</span>)</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>barcodes <span class="ot">&lt;-</span> vroom<span class="sc">::</span><span class="fu">vroom</span>(<span class="st">'scATAC/barcodes.tsv.gz'</span>) <span class="sc">|&gt;</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>features <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="st">'scATAC/features.tsv.gz'</span>)[<span class="sc">-</span><span class="dv">1</span>] <span class="sc">|&gt;</span> </span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">str_replace</span>(<span class="st">'_'</span>, <span class="st">':'</span>) <span class="sc">|&gt;</span> </span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">str_replace</span>(<span class="st">'_'</span>, <span class="st">'-'</span>) <span class="sc">|&gt;</span> </span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">GRanges</span>() <span class="sc">|&gt;</span> </span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">peak =</span> <span class="fu">as.character</span>(<span class="at">x =</span> _) <span class="sc">|&gt;</span> <span class="fu">str_replace</span>(<span class="st">':'</span>, <span class="st">'-'</span>))</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="do">## - Make sure that the row/colnames for the count matrix match those of </span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="co">#       barcodes and features</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(cnts) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(barcodes) <span class="ot">&lt;-</span> barcodes<span class="sc">$</span>Group_Barcode</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(cnts) <span class="ot">&lt;-</span> <span class="fu">names</span>(features) <span class="ot">&lt;-</span> features<span class="sc">$</span>peak</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="do">## - Get human gene annotations (hg19/GRCh37) to feed it into the future `SeuratChromatinAssay`</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> <span class="fu">GetGRangesFromEnsDb</span>(<span class="at">ensdb =</span> EnsDb.Hsapiens.v75<span class="sc">::</span>EnsDb.Hsapiens.v75)</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a><span class="fu">seqlevelsStyle</span>(annotations) <span class="ot">&lt;-</span> <span class="st">'UCSC'</span></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="fu">genome</span>(annotations) <span class="ot">&lt;-</span> <span class="st">"hg19"</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a><span class="do">## - Create Chromatin Assay</span></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>assay <span class="ot">&lt;-</span> Signac<span class="sc">::</span><span class="fu">CreateChromatinAssay</span>(</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>    <span class="at">counts =</span> cnts,</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>    <span class="at">ranges =</span> features,</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>    <span class="at">genome =</span> <span class="st">"hg19"</span>,</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">min.cells =</span> <span class="dv">10</span>, </span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">min.features =</span> <span class="dv">10</span>, </span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotation =</span> annotations</span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a>hemato <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">CreateSeuratObject</span>(</span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">counts =</span> assay,</span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay =</span> <span class="st">'ATAC'</span>,</span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">meta.data =</span> barcodes</span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>You will find that <code>CreateChromatinAssay()</code> function takes a long time to run. This is because it internally computes metrics from the count matrix. You can check the cell metadata to see which ones were calculated (on top of the ones which come from the <code>barcodes</code> data frame).</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">hemato</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">hemato</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">hemato</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">barcodes</span><span class="op">)</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="process-data" class="level4" data-number="15.1.2.2"><h4 data-number="15.1.2.2" class="anchored" data-anchor-id="process-data">
<span class="header-section-number">15.1.2.2</span> Process data</h4>
<ul>
<li><strong>Is the <code>fragments</code> file (which should exist for any scATACseq experiment) available for this experiment? Check the related GEO webpage.</strong></li>
<li><strong>For which analysis are the <code>fragments</code> essential, exactly?</strong></li>
<li><strong>Can we still perform normalization/clustering/annotation without them? And motif enrichment analysis?</strong></li>
</ul>
<p>Since we donâ€™t have the <code>fragments</code> file at hand, most of the QC steps are not available (e.g.&nbsp;<code>TSSEnrichment</code>, <code>NucleosomeSignal</code> or fragment size distribution).</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <code>fragments</code> file is actually available as raw data from GEO! Feel free to download it and perform further QC. Bare in mind that they are very large.</p>
</div>
</div>
<p>We do have other pre-computed metrics (e.g.&nbsp;<code>FRIP</code> and <code>depth</code>).</p>
<ul>
<li><strong>Check these metrics and filter the <code>Seurat</code> object (cells and features) as deemed appropriate.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Filter data </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">hemato</span><span class="op">$</span><span class="va">depth</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">hemato</span><span class="op">$</span><span class="va">nFeature_ATAC</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">hemato</span><span class="op">$</span><span class="va">nCount_ATAC</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">hemato</span><span class="op">$</span><span class="va">FRIP</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">hemato</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">hemato</span>, subset <span class="op">=</span> <span class="va">depth</span> <span class="op">&gt;</span> <span class="fl">5000</span> <span class="op">&amp;</span> <span class="va">depth</span> <span class="op">&lt;</span> <span class="fl">200000</span><span class="op">)</span> </span>
<span><span class="va">hemato</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">hemato</span>, subset <span class="op">=</span> <span class="va">nCount_ATAC</span> <span class="op">&gt;</span> <span class="fl">3000</span> <span class="op">&amp;</span> <span class="va">nCount_ATAC</span> <span class="op">&lt;</span> <span class="fl">100000</span><span class="op">)</span> </span>
<span><span class="va">hemato</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">hemato</span>, subset <span class="op">=</span> <span class="va">nFeature_ATAC</span> <span class="op">&gt;</span> <span class="fl">1000</span> <span class="op">&amp;</span> <span class="va">nFeature_ATAC</span> <span class="op">&lt;</span> <span class="fl">20000</span><span class="op">)</span> </span>
<span><span class="va">hemato</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">hemato</span>, subset <span class="op">=</span> <span class="va">FRIP</span> <span class="op">&gt;</span> <span class="fl">0.30</span> <span class="op">&amp;</span> <span class="va">FRIP</span> <span class="op">&lt;</span> <span class="fl">20000</span><span class="op">)</span> </span>
<span><span class="co">## - Remove peaks with low coverage</span></span>
<span><span class="va">hemato</span> <span class="op">&lt;-</span> <span class="va">hemato</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu">GetAssayData</span><span class="op">(</span><span class="va">hemato</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">10</span>, <span class="op">]</span></span>
<span><span class="va">hemato</span> <span class="op">&lt;-</span> <span class="va">hemato</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu">GetAssayData</span><span class="op">(</span><span class="va">hemato</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">10</span>, <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>Now that the dataset is filtered, normalize (by using TF-IDF approach) then further reduce the dimensionality for visualization purposes.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Normalize data </span></span>
<span><span class="va">hemato</span> <span class="op">&lt;-</span> <span class="fu">Signac</span><span class="fu">::</span><span class="fu">RunTFIDF</span><span class="op">(</span><span class="va">hemato</span><span class="op">)</span> </span>
<span><span class="co">## - Reduce dimensionality</span></span>
<span><span class="va">hemato</span> <span class="op">&lt;-</span> <span class="fu">Signac</span><span class="fu">::</span><span class="fu">FindTopFeatures</span><span class="op">(</span><span class="va">hemato</span>, min.cutoff <span class="op">=</span> <span class="st">'q50'</span><span class="op">)</span> </span>
<span><span class="va">hemato</span> <span class="op">&lt;-</span> <span class="fu">Signac</span><span class="fu">::</span><span class="fu">RunSVD</span><span class="op">(</span><span class="va">hemato</span><span class="op">)</span> </span>
<span><span class="co">## - Label clusters according to the original publication</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">'Cluster1'</span> <span class="op">=</span> <span class="st">'HSC'</span>,</span>
<span>    <span class="st">'Cluster2'</span> <span class="op">=</span> <span class="st">'MEP'</span>,</span>
<span>    <span class="st">'Cluster3'</span> <span class="op">=</span> <span class="st">'CMP'</span>,</span>
<span>    <span class="st">'Cluster4'</span> <span class="op">=</span> <span class="st">'LMPP'</span>,</span>
<span>    <span class="st">'Cluster5'</span> <span class="op">=</span> <span class="st">'CLP'</span>,</span>
<span>    <span class="st">'Cluster6'</span> <span class="op">=</span> <span class="st">'Pro-B'</span>,</span>
<span>    <span class="st">'Cluster7'</span> <span class="op">=</span> <span class="st">'Pre-B'</span>,</span>
<span>    <span class="st">'Cluster8'</span> <span class="op">=</span> <span class="st">'GMP'</span>,</span>
<span>    <span class="st">'Cluster9'</span> <span class="op">=</span> <span class="st">'MDP'</span>,</span>
<span>    <span class="st">'Cluster10'</span> <span class="op">=</span> <span class="st">'pDC'</span>,</span>
<span>    <span class="st">'Cluster11'</span> <span class="op">=</span> <span class="st">'cDC'</span>,</span>
<span>    <span class="st">'Cluster12'</span> <span class="op">=</span> <span class="st">'Mono.'</span>,</span>
<span>    <span class="st">'Cluster13'</span> <span class="op">=</span> <span class="st">'Mono.'</span>,</span>
<span>    <span class="st">'Cluster14'</span> <span class="op">=</span> <span class="st">'Nai. B'</span>,</span>
<span>    <span class="st">'Cluster15'</span> <span class="op">=</span> <span class="st">'Mem. B'</span>,</span>
<span>    <span class="st">'Cluster16'</span> <span class="op">=</span> <span class="st">'Plasma'</span>,</span>
<span>    <span class="st">'Cluster17'</span> <span class="op">=</span> <span class="st">'Baso.'</span>,</span>
<span>    <span class="st">'Cluster18'</span> <span class="op">=</span> <span class="st">'Imm. NK'</span>,</span>
<span>    <span class="st">'Cluster19'</span> <span class="op">=</span> <span class="st">'Mat. NK'</span>,</span>
<span>    <span class="st">'Cluster20'</span> <span class="op">=</span> <span class="st">'Mat. NK'</span>,</span>
<span>    <span class="st">'Cluster21'</span> <span class="op">=</span> <span class="st">'Nai. CD4'</span>,</span>
<span>    <span class="st">'Cluster22'</span> <span class="op">=</span> <span class="st">'Nai. CD4'</span>,</span>
<span>    <span class="st">'Cluster23'</span> <span class="op">=</span> <span class="st">'Nai. Treg'</span>,</span>
<span>    <span class="st">'Cluster24'</span> <span class="op">=</span> <span class="st">'Mem. CD4'</span>,</span>
<span>    <span class="st">'Cluster25'</span> <span class="op">=</span> <span class="st">'Treg'</span>,</span>
<span>    <span class="st">'Cluster26'</span> <span class="op">=</span> <span class="st">'Nai. CD8'</span>,</span>
<span>    <span class="st">'Cluster27'</span> <span class="op">=</span> <span class="st">'Nai. CD8'</span>,</span>
<span>    <span class="st">'Cluster28'</span> <span class="op">=</span> <span class="st">'Nai. CD8'</span>,</span>
<span>    <span class="st">'Cluster29'</span> <span class="op">=</span> <span class="st">'Mem. CD8'</span>,</span>
<span>    <span class="st">'Cluster30'</span> <span class="op">=</span> <span class="st">'Mem. CD8'</span>,</span>
<span>    <span class="st">'Cluster31'</span> <span class="op">=</span> <span class="st">'Gamm. Del. T'</span></span>
<span><span class="op">)</span></span>
<span><span class="va">hemato</span><span class="op">$</span><span class="va">renamed_clusters</span> <span class="op">&lt;-</span> <span class="va">clusters</span><span class="op">[</span><span class="va">hemato</span><span class="op">$</span><span class="va">Clusters</span><span class="op">]</span></span>
<span><span class="co">## - Visualize data </span></span>
<span><span class="va">hemato</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu">RunUMAP</span><span class="op">(</span><span class="va">hemato</span>, reduction <span class="op">=</span> <span class="st">'lsi'</span>, dims <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">hemato</span>, group.by <span class="op">=</span> <span class="st">'renamed_clusters'</span><span class="op">)</span> <span class="op">+</span> <span class="fu">coord_fixed</span><span class="op">(</span>ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><strong>What can you observe in the <code>UMAP</code> projection of the dataset? Comment on the separation of some cell types into different spatially-resolved clusters.</strong></li>
</ul></section></section></section><section id="chromvar-analysis" class="level2" data-number="15.2"><h2 data-number="15.2" class="anchored" data-anchor-id="chromvar-analysis">
<span class="header-section-number">15.2</span> chromVAR analysis</h2>
<section id="get-a-summarizedexperiment-of-scatacseq-counts-over-peaks" class="level3" data-number="15.2.1"><h3 data-number="15.2.1" class="anchored" data-anchor-id="get-a-summarizedexperiment-of-scatacseq-counts-over-peaks">
<span class="header-section-number">15.2.1</span> Get a SummarizedExperiment of scATACseq counts over peaks</h3>
<p><code>chromVAR</code> works with raw counts stored as a <code>RangedSummarizedExperiment</code>. By now, you should be able to easily extract the raw counts from a specific <code>Seurat</code> assay. Store it in a <code>RangedSummarizedExperiment</code> (~ equivalent to <code>Seurat</code> assay but in <code>Bioconductor</code>).</p>
<ul>
<li><strong>What are the dimensions of the <code>RangedSummarizedExperiment</code> generated? What are the features and what are the columns?</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="va">sumExp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">SummarizedExperiment</a></span><span class="op">(</span></span>
<span>    assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">'counts'</span> <span class="op">=</span> <span class="fu">GetAssayData</span><span class="op">(</span><span class="va">hemato</span>, assay <span class="op">=</span> <span class="st">'ATAC'</span>, slot <span class="op">=</span> <span class="st">'counts'</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    rowRanges <span class="op">=</span> <span class="fu">GetAssayData</span><span class="op">(</span><span class="va">hemato</span>, assay <span class="op">=</span> <span class="st">'ATAC'</span>, slot <span class="op">=</span> <span class="st">'ranges'</span><span class="op">)</span>, </span>
<span>    colData <span class="op">=</span> <span class="va">hemato</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">sumExp</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="add-gc-bias-to-peaks" class="level3" data-number="15.2.2"><h3 data-number="15.2.2" class="anchored" data-anchor-id="add-gc-bias-to-peaks">
<span class="header-section-number">15.2.2</span> Add GC bias to peaks</h3>
<p>This step is important to correct GC bias associated with each peak.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sumExp</span> <span class="op">&lt;-</span> <span class="fu">chromVAR</span><span class="fu">::</span><span class="fu">addGCBias</span><span class="op">(</span><span class="va">sumExp</span>, genome <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va">BSgenome.Hsapiens.UCSC.hg19</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="map-motifs-over-peaks" class="level3" data-number="15.2.3"><h3 data-number="15.2.3" class="anchored" data-anchor-id="map-motifs-over-peaks">
<span class="header-section-number">15.2.3</span> Map motifs over peaks</h3>
<p>Using <code>motifmatchr</code> package, one can map a list of motifs (e.g.&nbsp;from JASPAR database) over a <code>RangedSummarizedExperiment</code> (or anything which can be coerced to a <code>GRanges</code>, actually).</p>
<ul>
<li><strong>First, you need to import motifs from a public database in <code>R</code>. This can be done (among other ways) with the <code>TFBSTools</code> pacakge.</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Import motifs from JASPAR database</span></span>
<span><span class="va">motifs</span> <span class="op">&lt;-</span> <span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/getMatrixSet-methods.html">getMatrixSet</a></span><span class="op">(</span></span>
<span>    <span class="fu">JASPAR2020</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/JASPAR2020/man/JASPAR2020-class.html">JASPAR2020</a></span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>species <span class="op">=</span> <span class="fl">9606</span>, all_versions <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">motifs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">TFBSTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/XMatrix-class.html">name</a></span><span class="op">(</span><span class="va">motifs</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li><p><strong>Now map the subset of motifs of interest to the filtered peaks, using <code>motifmatchr::matchMotifs()</code>.</strong></p></li>
<li><p><strong>Read <code>matchMotifs()</code> documentation and run it on your <code>sumExp</code>.</strong></p></li>
<li><p><strong>What is the class of the generated object? Its dimensions?</strong></p></li>
<li><p><strong>What are the features and what are the columns?</strong></p></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">16</span>, progressbar <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">motif_mappings</span> <span class="op">&lt;-</span> <span class="fu">motifmatchr</span><span class="fu">::</span><span class="fu">matchMotifs</span><span class="op">(</span><span class="va">motifs</span>, <span class="va">sumExp</span>, genome <span class="op">=</span> <span class="fu">BSgenome.Hsapiens.UCSC.hg19</span><span class="fu">::</span><span class="va">BSgenome.Hsapiens.UCSC.hg19</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">motif_mappingss</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">motif_mappings</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>Because there are so many motifs and so many peaks, and in interest of time, we will perform <code>chromVAR</code> analysis on a subset of motifs only. You can manually filter the motifs list to motifs that may be relevant in our context (e.g.&nbsp;<code>EOMES</code>, or <code>GATA2</code>, â€¦).</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Filter to interesting motifs to keep analysis relatively quick</span></span>
<span><span class="co">#       Feel free to select any TF of interest!</span></span>
<span><span class="va">motif_mappings_sub</span> <span class="op">&lt;-</span> <span class="va">motif_mappings</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'EOMES'</span>, <span class="st">'GATA2'</span>, <span class="st">'IRF8'</span>, <span class="st">'EBF1'</span>, <span class="st">'SPI1'</span>, <span class="st">'TCF3'</span>, <span class="st">'CEBPA'</span><span class="op">)</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="search-for-motifs-with-high-deviation-of-mapping-compared-to-background" class="level3" data-number="15.2.4"><h3 data-number="15.2.4" class="anchored" data-anchor-id="search-for-motifs-with-high-deviation-of-mapping-compared-to-background">
<span class="header-section-number">15.2.4</span> Search for motifs with high deviation of mapping compared to background</h3>
<p><code>chromVAR</code>â€™s <code>computeDeviations()</code> function combines (1) the peak counts / cell (stored here in <code>sumExp</code>) and (2) the TF mapping over each peak (stored here in <code>motif_mappings</code>). to assess the mapping deviation for each TF over each cell compared to other cells.</p>
<ul>
<li><strong>Run <code>computeDeviations()</code> on the set of peaks and motifs.</strong></li>
<li><strong>What is the class of the generated object? Its dimensions?</strong></li>
<li><strong>What are the features and what are the columns?</strong></li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Find background signal</span></span>
<span><span class="va">bg</span> <span class="op">&lt;-</span> <span class="fu">chromVAR</span><span class="fu">::</span><span class="fu">getBackgroundPeaks</span><span class="op">(</span>object <span class="op">=</span> <span class="va">sumExp</span><span class="op">)</span></span>
<span><span class="co">## - For each motif, compute its mapping deviation over the filtered peaks</span></span>
<span><span class="va">motif_deviations</span> <span class="op">&lt;-</span> <span class="fu">chromVAR</span><span class="fu">::</span><span class="fu">computeDeviations</span><span class="op">(</span>object <span class="op">=</span> <span class="va">sumExp</span>, annotations <span class="op">=</span> <span class="va">motif_mappings_sub</span>, background_peaks <span class="op">=</span> <span class="va">bg</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="check-tf-motif-enrichment-in-different-cell-types" class="level3" data-number="15.2.5"><h3 data-number="15.2.5" class="anchored" data-anchor-id="check-tf-motif-enrichment-in-different-cell-types">
<span class="header-section-number">15.2.5</span> Check TF motif enrichment in different cell types</h3>
<p>The <code>motif_deviations</code> object can be added as a new assay to the <code>hemato</code> <code>Seurat</code> object. This way, one can rely on <code>Seurat</code>-based plotting functions to plot cells in their preferred dimensional space, and color them using motif deviation scores computed with <code>chromVAR</code>.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hemato</span><span class="op">[[</span><span class="st">'MOTIF'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu">CreateAssayObject</span><span class="op">(</span>counts <span class="op">=</span> <span class="fu">chromVAR</span><span class="fu">::</span><span class="fu">deviationScores</span><span class="op">(</span><span class="va">motif_deviations</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">hemato</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'MOTIF'</span></span>
<span><span class="va">list_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">hemato</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">motif</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">FeaturePlot</span><span class="op">(</span><span class="va">hemato</span>, features <span class="op">=</span> <span class="va">motif</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>        <span class="fu">scale_colour_gradientn</span><span class="op">(</span></span>
<span>            colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'#190886'</span>, <span class="st">'#6F07F8'</span>, <span class="st">'#F954A5'</span>, <span class="st">'#FF9D66'</span>, <span class="st">'#edf118'</span><span class="op">)</span>,</span>
<span>            limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span>, oob <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/oob.html">squish</a></span></span>
<span>        <span class="op">)</span> <span class="op">+</span> </span>
<span>        <span class="fu">theme</span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">list_p</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">list_p</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">hemato</span>, group.by <span class="op">=</span> <span class="st">'renamed_clusters'</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">list_p</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section></section><section id="compute-gene-activity-scores" class="level2" data-number="15.3"><h2 data-number="15.3" class="anchored" data-anchor-id="compute-gene-activity-scores">
<span class="header-section-number">15.3</span> Compute gene activity scores</h2>
<p>Here again, the absence of <code>fragments</code> complicates the computation of gene activity scores. Indeed, Signacâ€™s <code>GeneActivity()</code> function does require scATACseq fragment information.</p>
<ul>
<li><strong>Without actual fragments, can you think of a way to still compute a â€˜gene activityâ€™-like score, approximating ~ the sum of scATACseq fragments overlapping a gene?</strong></li>
</ul>
<p>One way to do this is to:</p>
<ol type="1">
<li>Extend gene annotations to cover 2000bp upstream of TSS</li>
<li>For each gene, link it to all the peaks overlapping it to create a gene &lt;-&gt; peak relationship table</li>
<li>Create a joint data table containing [ gene | peak | cell | count ]</li>
<li>Group the data table by <code>cell + gene</code> and sum up the peak counts for each pair</li>
<li>Transform this count table into a SparseMatrix</li>
<li>Create the new <code>Seurat</code> assay containing â€œgene activityâ€-like scores and normalize it</li>
</ol>
<section id="manually-calculate-gene-activity-scores" class="level3" data-number="15.3.1"><h3 data-number="15.3.1" class="anchored" data-anchor-id="manually-calculate-gene-activity-scores">
<span class="header-section-number">15.3.1</span> Manually calculate gene activity scores</h3>
<ul>
<li>Step 1:</li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="do">## - Extend all the genes in hg19 so that it covers 1000 bp upstream of the genebody </span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>hg_genes_extended <span class="ot">&lt;-</span> ensembldb<span class="sc">::</span><span class="fu">genes</span>(EnsDb.Hsapiens.v75<span class="sc">::</span>EnsDb.Hsapiens.v75) <span class="sc">|&gt;</span> </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(gene_biotype <span class="sc">==</span> <span class="st">'protein_coding'</span>) <span class="sc">|&gt;</span> </span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">keepStandardChromosomes</span>(<span class="at">pruning.mode =</span> <span class="st">'coarse'</span>) <span class="sc">|&gt;</span> </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(seqnames <span class="sc">!=</span> <span class="st">'MT'</span>) <span class="sc">|&gt;</span> </span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">resize</span>(<span class="at">width =</span> <span class="fu">width</span>(<span class="at">x =</span> _) <span class="sc">+</span> <span class="dv">2000</span>, <span class="at">fix =</span> <span class="st">'start'</span>) <span class="sc">|&gt;</span> </span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">shift</span>(<span class="dv">1000</span> <span class="sc">*</span> <span class="fu">ifelse</span>(<span class="fu">strand</span>(<span class="at">x =</span> _) <span class="sc">==</span> <span class="st">'+'</span>, <span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>)) <span class="sc">|&gt;</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">id =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="at">x =</span> _)) <span class="sc">|&gt;</span> </span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">gene_name =</span> <span class="fu">make.unique</span>(gene_name))</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="fu">seqlevelsStyle</span>(hg_genes_extended) <span class="ot">&lt;-</span> <span class="st">'UCSC'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li>Step 2:</li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Find overlaps between all the scATAC peaks and the extended genes</span></span>
<span><span class="va">ov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">findOverlaps</a></span><span class="op">(</span>query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html">granges</a></span><span class="op">(</span><span class="va">hemato</span><span class="op">)</span>, subject <span class="op">=</span> <span class="va">hg_genes_extended</span><span class="op">)</span></span>
<span><span class="co">## - For each peak-gene pair, add peak name and gene name</span></span>
<span><span class="va">ov</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">ov</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>peaknb <span class="op">=</span> <span class="va">queryHits</span>, genenb <span class="op">=</span> <span class="va">subjectHits</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">mutate</span><span class="op">(</span>gene_name <span class="op">=</span> <span class="va">hg_genes_extended</span><span class="op">$</span><span class="va">gene_name</span><span class="op">[</span><span class="va">genenb</span><span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">mutate</span><span class="op">(</span>row <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html">granges</a></span><span class="op">(</span><span class="va">hemato</span><span class="op">)</span><span class="op">$</span><span class="va">peak</span><span class="op">[</span><span class="va">peaknb</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li>Step 3:</li>
</ul>
<p>This step requires to transform the original <code>counts</code> SparseMatrix into a tidy <code>data.table</code>, to enable fast <code>left_join</code> between (1) peak counts / cell and (2) gene &lt;-&gt; peak relationship table (<code>ov</code>)</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Extract counts for the gene-associated peaks</span></span>
<span><span class="va">cnts</span> <span class="op">&lt;-</span> <span class="fu">GetAssayData</span><span class="op">(</span><span class="va">hemato</span>, <span class="st">'counts'</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">ov</span><span class="op">$</span><span class="va">peaknb</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co">## - Transform the cnts SparseMatrix into a tidy dense long data.frame</span></span>
<span><span class="va">tidy_cnts</span> <span class="op">&lt;-</span> <span class="fu">broom</span><span class="fu">::</span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html">tidy</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span> </span>
<span><span class="co">## - Transform the dense long data.frame into a data.table for faster computation</span></span>
<span><span class="va">tidy_cnts</span> <span class="op">&lt;-</span> <span class="fu">dtplyr</span><span class="fu">::</span><span class="fu"><a href="https://dtplyr.tidyverse.org/reference/lazy_dt.html">lazy_dt</a></span><span class="op">(</span><span class="va">tidy_cnts</span><span class="op">)</span></span>
<span><span class="co">## - For each peak/cell count value, associate the corresponding gene</span></span>
<span><span class="va">joint_data</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">tidy_cnts</span>, <span class="va">ov</span>, by <span class="op">=</span> <span class="st">'row'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">joint_data</span><span class="op">)</span></span>
<span><span class="co">## WATCH OUT: You cannot see what is stored into `joint_data` as it </span></span>
<span><span class="co">## is a deferred data.table, i.e. it will only execute once you pipe it to </span></span>
<span><span class="co">## `as_tibble()`. Once you pipe it, the computation starts and takes a lot of time.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li>Step 4:</li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Group the joint_table by gene + cell and sum up all the counts</span></span>
<span><span class="va">gene_averaged_cnts</span> <span class="op">&lt;-</span> <span class="fu">group_by</span><span class="op">(</span><span class="va">joint_data</span>, <span class="va">column</span>, <span class="va">gene_name</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">summarize</span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li>Step 5:</li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - To transform the data.table into SparseMatrix, it needs to be numerical</span></span>
<span><span class="va">gene_averaged_cnts_num</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.matrix.html">data.matrix</a></span><span class="op">(</span><span class="va">gene_averaged_cnts</span><span class="op">)</span></span>
<span><span class="co">## - Transform into sparse matrix </span></span>
<span><span class="va">gene_averaged_cnts_sparse</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/sparseMatrix.html">sparseMatrix</a></span><span class="op">(</span></span>
<span>    i <span class="op">=</span> <span class="va">gene_averaged_cnts_num</span><span class="op">[</span>, <span class="st">'gene_name'</span><span class="op">]</span>,</span>
<span>    j <span class="op">=</span> <span class="va">gene_averaged_cnts_num</span><span class="op">[</span>, <span class="st">'column'</span><span class="op">]</span>,</span>
<span>    x <span class="op">=</span> <span class="va">gene_averaged_cnts_num</span><span class="op">[</span>, <span class="st">'n'</span><span class="op">]</span>,</span>
<span><span class="op">)</span></span>
<span><span class="co">## - Add back the actual cell and genes names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">gene_averaged_cnts_sparse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">gene_averaged_cnts</span><span class="op">$</span><span class="va">column</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gene_averaged_cnts_sparse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">gene_averaged_cnts</span><span class="op">$</span><span class="va">gene_name</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<ul>
<li>Step 6:</li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Create new assay in hemato</span></span>
<span><span class="va">hemato</span><span class="op">[[</span><span class="st">'GENE'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">CreateAssayObject</span><span class="op">(</span>counts <span class="op">=</span> <span class="va">gene_averaged_cnts_sparse</span><span class="op">)</span></span>
<span><span class="co">## - Normalize this new assay with SCTransform</span></span>
<span><span class="va">hemato</span> <span class="op">&lt;-</span> <span class="fu">SCTransform</span><span class="op">(</span><span class="va">hemato</span>, assay <span class="op">=</span> <span class="st">'GENE'</span><span class="op">)</span></span>
<span><span class="co">## - Save the resulting object!</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">hemato</span>, <span class="st">'scATAC/hemato.rds'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>Now the <code>Seurat</code> object contains 3 assays, the first one corresponding to the original scATACseq counts, and the second one containing gene activity pseudocounts, and the third one containing <code>SCT</code>-normalized data. One can now perform â€œgene differential expressionâ€-like analysis using the <code>SCT</code> assay!</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Reduce dimensionality to visualize cells in 2D</span></span>
<span><span class="va">hemato</span> <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">hemato</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">RunUMAP</span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">'pca'</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span></span>
<span><span class="co">## - Find gene markers for Basophils</span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">hemato</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SCT"</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">FeaturePlot</span><span class="op">(</span><span class="va">hemato</span>, features <span class="op">=</span> <span class="st">'CD14'</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span> <span class="co"># Or NCR1 (NKp46), CD8A, CD14, MS4A1</span></span>
<span><span class="fu">Idents</span><span class="op">(</span><span class="va">hemato</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">hemato</span><span class="op">$</span><span class="va">renamed_clusters</span></span>
<span><span class="va">basophils_genes</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu">FindMarkers</span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">hemato</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">'GENE'</span>, </span>
<span>    ident.1 <span class="op">=</span> <span class="st">'Baso.'</span>,</span>
<span>    ident.2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="fu">Idents</span><span class="op">(</span><span class="va">hemato</span><span class="op">)</span><span class="op">[</span><span class="fu">Idents</span><span class="op">(</span><span class="va">hemato</span><span class="op">)</span> <span class="op">!=</span> <span class="st">'Baso.'</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    min.pct <span class="op">=</span> <span class="fl">0.15</span>, </span>
<span>    only.pos <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="compare-gene-activity-and-tf-activity" class="level3" data-number="15.3.2"><h3 data-number="15.3.2" class="anchored" data-anchor-id="compare-gene-activity-and-tf-activity">
<span class="header-section-number">15.3.2</span> Compare gene activity and TF activity</h3>
<p>Plot side-by-side:</p>
<ul>
<li>A UMAP projection showing where cells along the B lineage are located</li>
<li>A UMAP projection showing <code>EBF1</code> gene activity per cell</li>
<li>A UMAP projection showing <code>EBF1</code> motif enrichment (i.e.&nbsp;TF activity) per cell</li>
</ul>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">'HSC'</span>, </span>
<span>    <span class="st">'MPP'</span>,</span>
<span>    <span class="st">'LMPP'</span>, </span>
<span>    <span class="st">'CLP'</span>, </span>
<span>    <span class="st">'Pro-B'</span>, </span>
<span>    <span class="st">'Pre-B'</span>, </span>
<span>    <span class="st">'Nai. B'</span>, </span>
<span>    <span class="st">'Mem. B'</span>, </span>
<span>    <span class="st">'Plasma'</span></span>
<span><span class="op">)</span></span>
<span><span class="va">hemato</span><span class="op">$</span><span class="va">b_cells_annots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">hemato</span><span class="op">$</span><span class="va">renamed_clusters</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">labels</span>, <span class="va">hemato</span><span class="op">$</span><span class="va">renamed_clusters</span>, <span class="cn">NA</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">hemato</span>, group.by <span class="op">=</span> <span class="st">'b_cells_annots'</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_colour_discrete</span><span class="op">(</span>na.value <span class="op">=</span> <span class="st">'#aaaaaa'</span><span class="op">)</span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">hemato</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'SCT'</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">FeaturePlot</span><span class="op">(</span><span class="va">hemato</span>, features <span class="op">=</span> <span class="st">'EBF1'</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>        <span class="fu">scale_colour_gradientn</span><span class="op">(</span></span>
<span>            colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'#dadad8'</span>, <span class="st">'#f1eca3'</span>, <span class="st">'#e0c067'</span>, <span class="st">'#db8035'</span>, <span class="st">'#a3160c'</span><span class="op">)</span>,</span>
<span>            limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span>, oob <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/oob.html">squish</a></span></span>
<span>        <span class="op">)</span> <span class="op">+</span> </span>
<span>        <span class="fu">theme</span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">hemato</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'MOTIF'</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu">FeaturePlot</span><span class="op">(</span><span class="va">hemato</span>, features <span class="op">=</span> <span class="st">'EBF1'</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>        <span class="fu">scale_colour_gradientn</span><span class="op">(</span></span>
<span>            colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'#190886'</span>, <span class="st">'#6F07F8'</span>, <span class="st">'#F954A5'</span>, <span class="st">'#FF9D66'</span>, <span class="st">'#edf118'</span><span class="op">)</span>,</span>
<span>            limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span>, oob <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/oob.html">squish</a></span></span>
<span>        <span class="op">)</span> <span class="op">+</span> </span>
<span>        <span class="fu">theme</span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, <span class="va">p3</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>


</section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../content/day4/Lecture7_ATAC.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Lecture 7 - Advances in single-cell genomics: the epigenome</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../content/day5/Lecture8_spatial-transcriptomics.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Lecture 8 - Advances in single-cell genomics: spatial transcriptomics</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">Single-cell RNAseq analysis with R/Bioconductor |<br>
J. Serizay O. Ashenberg</div>   
    <div class="nav-footer-right">This book was built with <a href="https://quarto.org/">Quarto</a>.</div>
  </div>
</footer>


</body></html>